// eslint-plugin-astro/lib/index.mjs
import Module, { createRequire } from 'module';
import path4 from 'path';
import { getSourceCode as getSourceCode$1, getFilename as getFilename$1, getCwd as getCwd$1 } from 'eslint-compat-utils';
import { AST_NODE_TYPES } from '@typescript-eslint/types';
import { ReferenceTracker, READ, getPropertyName, isOpeningParenToken, isSemicolonToken, getStaticValue, isParenthesized, isClosingParenToken, isCommaToken, isClosingBraceToken } from '@eslint-community/eslint-utils';
import postcss from 'postcss';
import { decode } from '@jridgewell/sourcemap-codec';
import parser from 'postcss-selector-parser';
import * as parser2 from 'astro-eslint-parser';
import { parseTemplate, traverseNodes } from 'astro-eslint-parser';
import globals from 'globals';

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all2) => {
  for (var name2 in all2)
    __defProp(target, name2, { get: all2[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var hasTypescriptEslintParser, tsESLintParser;
var init_has_typescript_eslint_parser = __esm({
  "src/configs/has-typescript-eslint-parser.ts"() {
    hasTypescriptEslintParser = false;
    tsESLintParser = null;
    try {
      const cwd = process.cwd();
      const relativeTo = path4.join(cwd, "__placeholder__.js");
      if (tsESLintParser = createRequire(relativeTo)("@typescript-eslint/parser"))
        hasTypescriptEslintParser = true;
    } catch {
    }
  }
});

// src/environments/index.ts
var environments;
var init_environments = __esm({
  "src/environments/index.ts"() {
    environments = {
      astro: {
        globals: {
          // Astro object
          Astro: false,
          // JSX Fragment
          Fragment: false
        }
      }
    };
  }
});

// src/utils/index.ts
function createRule(ruleName, rule) {
  return {
    meta: {
      ...rule.meta,
      docs: {
        available: () => true,
        ...rule.meta.docs,
        url: `https://ota-meshi.github.io/eslint-plugin-astro/rules/${ruleName}/`,
        ruleId: `astro/${ruleName}`,
        ruleName
      }
    },
    create: rule.create
  };
}
var init_utils = __esm({
  "src/utils/index.ts"() {
  }
});
function getSourceCode(context) {
  return getSourceCode$1(context);
}
function getFilename(context) {
  return getFilename$1(context);
}
function getCwd(context) {
  return getCwd$1(context);
}
var init_compat = __esm({
  "src/utils/compat.ts"() {
  }
});
function getAttributeName(node) {
  if (node.type === "JSXSpreadAttribute") {
    return null;
  }
  const { name: name2 } = node;
  return getName(name2);
}
function getElementName(node) {
  const nameNode = node.openingElement.name;
  return getName(nameNode);
}
function findAttribute(node, name2) {
  const openingElement = node.openingElement;
  for (const attr of openingElement.attributes) {
    if (attr.type === "JSXSpreadAttribute") {
      continue;
    }
    if (getAttributeName(attr) === name2) {
      return attr;
    }
  }
  return null;
}
function getSpreadAttributes(node) {
  const openingElement = node.openingElement;
  return openingElement.attributes.filter(
    (attr) => attr.type === "JSXSpreadAttribute"
  );
}
function getStaticAttributeStringValue(node, context) {
  const value = getStaticAttributeValue(node, context);
  if (!value) {
    return null;
  }
  return value.value != null ? String(value.value) : value.value;
}
function getStaticAttributeValue(node, context) {
  if (node.value?.type === AST_NODE_TYPES.Literal) {
    return { value: node.value.value };
  }
  if (context && node.value?.type === "JSXExpressionContainer" && node.value.expression.type !== "JSXEmptyExpression") {
    const sourceCode = getSourceCode(context);
    const staticValue = getStaticValue(
      node.value.expression,
      sourceCode.scopeManager.globalScope
    );
    if (staticValue != null) {
      return staticValue;
    }
  }
  return null;
}
function isStringCallExpression(node) {
  if (node.type === AST_NODE_TYPES.CallExpression) {
    return node.callee.type === AST_NODE_TYPES.Identifier && node.callee.name === "String";
  }
  return false;
}
function isStringLiteral(node) {
  return node.type === AST_NODE_TYPES.Literal && typeof node.value === "string";
}
function extractConcatExpressions(node, sourceCode) {
  if (node.operator !== "+") {
    return null;
  }
  const leftResult = processLeft(node.left);
  if (leftResult == null) {
    return null;
  }
  return [...leftResult, node.right];
  function processLeft(expr) {
    if (expr.type === AST_NODE_TYPES.BinaryExpression) {
      if (!isParenthesized(expr, sourceCode) && expr.operator !== "*" && expr.operator !== "/") {
        return extractConcatExpressions(expr, sourceCode);
      }
    }
    return [expr];
  }
}
function getStringIfConstant(node) {
  if (node.type === "Literal") {
    if (typeof node.value === "string") return node.value;
  } else if (node.type === "TemplateLiteral") {
    let str = "";
    const quasis = [...node.quasis];
    const expressions = [...node.expressions];
    let quasi, expr;
    while (quasi = quasis.shift()) {
      str += quasi.value.cooked;
      expr = expressions.shift();
      if (expr) {
        const exprStr = getStringIfConstant(expr);
        if (exprStr == null) {
          return null;
        }
        str += exprStr;
      }
    }
    return str;
  } else if (node.type === "BinaryExpression") {
    if (node.operator === "+") {
      const left = getStringIfConstant(node.left);
      if (left == null) {
        return null;
      }
      const right = getStringIfConstant(node.right);
      if (right == null) {
        return null;
      }
      return left + right;
    }
  }
  return null;
}
function needParentheses(node, kind) {
  if (node.type === "ArrowFunctionExpression" || node.type === "AssignmentExpression" || node.type === "BinaryExpression" || node.type === "ConditionalExpression" || node.type === "LogicalExpression" || node.type === "SequenceExpression" || node.type === "UnaryExpression" || node.type === "UpdateExpression")
    return true;
  return false;
}
function getParenthesizedTokens(node, sourceCode) {
  let lastLeft = sourceCode.getFirstToken(node);
  let lastRight = sourceCode.getLastToken(node);
  let maybeLeftParen, maybeRightParen;
  while ((maybeLeftParen = sourceCode.getTokenBefore(lastLeft)) && (maybeRightParen = sourceCode.getTokenAfter(lastRight)) && isOpeningParenToken(maybeLeftParen) && isClosingParenToken(maybeRightParen) && // Avoid false positive such as `if (a) {}`
  maybeLeftParen !== getParentSyntaxParen(node, sourceCode)) {
    lastLeft = maybeLeftParen;
    lastRight = maybeRightParen;
    maybeLeftParen = sourceCode.getTokenBefore(lastLeft);
    maybeRightParen = sourceCode.getTokenAfter(lastRight);
  }
  return { left: lastLeft, right: lastRight };
}
function getParenthesizedRange(node, sourceCode) {
  const { left, right } = getParenthesizedTokens(node, sourceCode);
  return [left.range[0], right.range[1]];
}
function getParentSyntaxParen(node, sourceCode) {
  const parent = node.parent;
  switch (parent.type) {
    case "CallExpression":
    case "NewExpression":
      if (parent.arguments.length === 1 && parent.arguments[0] === node) {
        return sourceCode.getTokenAfter(parent.callee, {
          includeComments: false,
          filter: isOpeningParenToken
        });
      }
      return null;
    case "DoWhileStatement":
      if (parent.test === node) {
        return sourceCode.getTokenAfter(parent.body, {
          includeComments: false,
          filter: isOpeningParenToken
        });
      }
      return null;
    case "IfStatement":
    case "WhileStatement":
      if (parent.test === node) {
        return sourceCode.getFirstToken(parent, {
          includeComments: false,
          skip: 1
        });
      }
      return null;
    case "ImportExpression":
      if (parent.source === node) {
        return sourceCode.getFirstToken(parent, {
          includeComments: false,
          skip: 1
        });
      }
      return null;
    case "SwitchStatement":
      if (parent.discriminant === node) {
        return sourceCode.getFirstToken(parent, {
          includeComments: false,
          skip: 1
        });
      }
      return null;
    case "WithStatement":
      if (parent.object === node) {
        return sourceCode.getFirstToken(parent, {
          includeComments: false,
          skip: 1
        });
      }
      return null;
    default:
      return null;
  }
}
function getName(nameNode) {
  if (nameNode.type === "JSXIdentifier") {
    return nameNode.name;
  }
  if (nameNode.type === "JSXNamespacedName") {
    return `${nameNode.namespace.name}:${nameNode.name.name}`;
  }
  if (nameNode.type === "JSXMemberExpression") {
    return `${getName(nameNode.object)}.${nameNode.property.name}`;
  }
  return null;
}
function isTokenOnSameLine(left, right) {
  return left?.loc?.end.line === right?.loc?.start.line;
}
function getNextLocation(sourceCode, { column, line }) {
  if (column < sourceCode.lines[line - 1].length) {
    return {
      column: column + 1,
      line
    };
  }
  if (line < sourceCode.lines.length) {
    return {
      column: 0,
      line: line + 1
    };
  }
  return null;
}
function getUpperFunction(node) {
  for (let currentNode = node; currentNode; currentNode = currentNode.parent) {
    if (anyFunctionPattern.test(currentNode.type)) return currentNode;
  }
  return null;
}
var anyFunctionPattern;
var init_ast_utils = __esm({
  "src/utils/ast-utils.ts"() {
    init_compat();
    anyFunctionPattern = /^(?:Function(?:Declaration|Expression)|ArrowFunctionExpression)$/u;
  }
});

// src/rules/missing-client-only-directive-value.ts
var missing_client_only_directive_value_default;
var init_missing_client_only_directive_value = __esm({
  "src/rules/missing-client-only-directive-value.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    missing_client_only_directive_value_default = createRule("missing-client-only-directive-value", {
      meta: {
        docs: {
          description: "the client:only directive is missing the correct component's framework value",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          missingValue: "`client:only` directive is missing a value"
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        function verifyDirectiveValue(attr) {
          const directiveName = getAttributeName(attr);
          if (directiveName !== "client:only") return;
          const directiveValue = getStaticAttributeValue(attr, context);
          if (directiveValue !== null) return;
          context.report({
            node: attr.name,
            messageId: "missingValue"
          });
        }
        return {
          JSXAttribute: verifyDirectiveValue,
          AstroTemplateLiteralAttribute: verifyDirectiveValue
        };
      }
    });
  }
});

// src/rules/no-conflict-set-directives.ts
var no_conflict_set_directives_default;
var init_no_conflict_set_directives = __esm({
  "src/rules/no-conflict-set-directives.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    no_conflict_set_directives_default = createRule("no-conflict-set-directives", {
      meta: {
        docs: {
          description: "disallow conflicting set directives and child contents",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          conflict: "{{name}} conflicts with {{conflictTargets}}."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          JSXElement(node) {
            const reportData = [];
            for (const attr of node.openingElement.attributes) {
              const directiveName = getAttributeName(attr);
              if (directiveName === "set:text" || directiveName === "set:html") {
                reportData.push({
                  loc: attr.loc,
                  name: `'${directiveName}'`
                });
              }
            }
            if (reportData.length) {
              const targetChildren = node.children.filter((child) => {
                if (child.type === "AstroHTMLComment") {
                  return false;
                }
                if (child.type === "JSXText" || child.type === "AstroRawText") {
                  return Boolean(child.value.trim());
                }
                return true;
              }).map((child) => {
                if (child.type === "JSXText" || child.type === "AstroRawText") {
                  const leadingSpaces = /^\s*/.exec(child.value)[0];
                  const trailingSpaces = /\s*$/.exec(child.value)[0];
                  return {
                    loc: {
                      start: sourceCode.getLocFromIndex(
                        child.range[0] + leadingSpaces.length
                      ),
                      end: sourceCode.getLocFromIndex(
                        child.range[1] - trailingSpaces.length
                      )
                    }
                  };
                }
                return child;
              });
              if (targetChildren.length) {
                reportData.push({
                  loc: {
                    start: targetChildren[0].loc.start,
                    end: targetChildren[targetChildren.length - 1].loc.end
                  },
                  name: "child contents"
                });
              }
            }
            if (reportData.length >= 2) {
              for (const data of reportData) {
                const conflictTargets = reportData.filter((d) => d !== data).map((d) => d.name);
                context.report({
                  loc: data.loc,
                  messageId: "conflict",
                  data: {
                    name: data.name,
                    conflictTargets: [
                      conflictTargets.slice(0, -1).join(", "),
                      conflictTargets.slice(-1)[0]
                    ].filter(Boolean).join(", and ")
                  }
                });
              }
            }
          }
        };
      }
    });
  }
});
var no_deprecated_astro_canonicalurl_default;
var init_no_deprecated_astro_canonicalurl = __esm({
  "src/rules/no-deprecated-astro-canonicalurl.ts"() {
    init_utils();
    init_compat();
    no_deprecated_astro_canonicalurl_default = createRule("no-deprecated-astro-canonicalurl", {
      meta: {
        docs: {
          description: "disallow using deprecated `Astro.canonicalURL`",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          deprecated: "'Astro.canonicalURL' is deprecated. Use 'Astro.url' helper instead."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          "Program:exit"(node) {
            const tracker = new ReferenceTracker(sourceCode.getScope(node));
            for (const { node: node2, path: path5 } of tracker.iterateGlobalReferences({
              Astro: {
                canonicalURL: { [READ]: true }
              }
            })) {
              context.report({
                node: node2,
                messageId: "deprecated",
                data: { name: path5.join(".") }
              });
            }
          }
        };
      }
    });
  }
});
var no_deprecated_astro_fetchcontent_default;
var init_no_deprecated_astro_fetchcontent = __esm({
  "src/rules/no-deprecated-astro-fetchcontent.ts"() {
    init_utils();
    init_compat();
    no_deprecated_astro_fetchcontent_default = createRule("no-deprecated-astro-fetchcontent", {
      meta: {
        docs: {
          description: "disallow using deprecated `Astro.fetchContent()`",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          deprecated: "'Astro.fetchContent()' is deprecated. Use 'Astro.glob()' instead."
        },
        type: "problem",
        fixable: "code"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          "Program:exit"(node) {
            const tracker = new ReferenceTracker(sourceCode.getScope(node));
            for (const { node: node2, path: path5 } of tracker.iterateGlobalReferences({
              Astro: {
                fetchContent: { [READ]: true }
              }
            })) {
              context.report({
                node: node2,
                messageId: "deprecated",
                data: { name: path5.join(".") },
                fix(fixer) {
                  if (node2.type !== "MemberExpression" || node2.computed) {
                    return null;
                  }
                  return fixer.replaceText(node2.property, "glob");
                }
              });
            }
          }
        };
      }
    });
  }
});
var no_deprecated_astro_resolve_default;
var init_no_deprecated_astro_resolve = __esm({
  "src/rules/no-deprecated-astro-resolve.ts"() {
    init_utils();
    init_compat();
    no_deprecated_astro_resolve_default = createRule("no-deprecated-astro-resolve", {
      meta: {
        docs: {
          description: "disallow using deprecated `Astro.resolve()`",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          deprecated: "'Astro.resolve()' is deprecated."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          "Program:exit"(node) {
            const tracker = new ReferenceTracker(sourceCode.getScope(node));
            for (const { node: node2, path: path5 } of tracker.iterateGlobalReferences({
              Astro: {
                resolve: { [READ]: true }
              }
            })) {
              context.report({
                node: node2,
                messageId: "deprecated",
                data: { name: path5.join(".") }
              });
            }
          }
        };
      }
    });
  }
});

// src/rules/no-deprecated-getentrybyslug.ts
var no_deprecated_getentrybyslug_default;
var init_no_deprecated_getentrybyslug = __esm({
  "src/rules/no-deprecated-getentrybyslug.ts"() {
    init_utils();
    init_compat();
    no_deprecated_getentrybyslug_default = createRule("no-deprecated-getentrybyslug", {
      meta: {
        docs: {
          description: "disallow using deprecated `getEntryBySlug()`",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          deprecated: "'getEntryBySlug()' is deprecated. Use 'getEntry()' instead."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          ImportSpecifier(node) {
            if (node.imported.type === "Identifier" && node.imported.name === "getEntryBySlug" && node.parent?.type === "ImportDeclaration" && node.parent.source.value === "astro:content") {
              context.report({
                node,
                messageId: "deprecated"
              });
            }
          }
        };
      }
    });
  }
});

// src/rules/no-exports-from-components.ts
var ALLOWED_EXPORTS, no_exports_from_components_default;
var init_no_exports_from_components = __esm({
  "src/rules/no-exports-from-components.ts"() {
    init_utils();
    init_compat();
    ALLOWED_EXPORTS = /* @__PURE__ */ new Set(["getStaticPaths", "partial", "prerender"]);
    no_exports_from_components_default = createRule("no-exports-from-components", {
      meta: {
        docs: {
          description: "disallow value export",
          category: "Possible Errors",
          // TODO: Switch to recommended: true, in next major version
          recommended: false
        },
        schema: [],
        messages: {
          disallowExport: "Exporting values from components is not allowed."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        function verifyDeclaration(node) {
          if (!node) return;
          if (node.type.startsWith("TS") && !node.type.endsWith("Expression")) {
            return;
          }
          if (node.type === "FunctionDeclaration" && node.id && ALLOWED_EXPORTS.has(node.id.name) || node.type === "VariableDeclaration" && node.declarations.every(
            (decl) => decl.id.type === "Identifier" && ALLOWED_EXPORTS.has(decl.id.name)
          )) {
            return;
          }
          context.report({
            node,
            messageId: "disallowExport"
          });
        }
        return {
          ExportAllDeclaration(node) {
            if (node.exportKind === "type") return;
            context.report({
              node,
              messageId: "disallowExport"
            });
          },
          ExportDefaultDeclaration(node) {
            if (node.exportKind === "type") return;
            verifyDeclaration(node.declaration);
          },
          ExportNamedDeclaration(node) {
            if (node.exportKind === "type") return;
            verifyDeclaration(node.declaration);
            for (const spec of node.specifiers) {
              if (spec.exportKind === "type" || spec.exported.type !== "Identifier")
                continue;
              if (ALLOWED_EXPORTS.has(spec.exported.name)) {
                continue;
              }
              context.report({
                node: spec,
                messageId: "disallowExport"
              });
            }
          }
        };
      }
    });
  }
});

// src/rules/no-set-html-directive.ts
var no_set_html_directive_default;
var init_no_set_html_directive = __esm({
  "src/rules/no-set-html-directive.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    no_set_html_directive_default = createRule("no-set-html-directive", {
      meta: {
        docs: {
          description: "disallow use of `set:html` to prevent XSS attack",
          category: "Security Vulnerability",
          recommended: false
        },
        schema: [],
        messages: {
          unexpected: "`set:html` can lead to XSS attack."
        },
        type: "suggestion"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        function verifyName(attr) {
          if (getAttributeName(attr) !== "set:html") {
            return;
          }
          context.report({
            node: attr.name,
            messageId: "unexpected"
          });
        }
        return {
          JSXAttribute: verifyName,
          AstroTemplateLiteralAttribute: verifyName
        };
      }
    });
  }
});

// src/rules/no-set-text-directive.ts
var no_set_text_directive_default;
var init_no_set_text_directive = __esm({
  "src/rules/no-set-text-directive.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    no_set_text_directive_default = createRule("no-set-text-directive", {
      meta: {
        docs: {
          description: "disallow use of `set:text`",
          category: "Best Practices",
          recommended: false
        },
        schema: [],
        messages: {
          disallow: "Don't use `set:text`."
        },
        type: "suggestion",
        fixable: "code"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        function verifyName(attr) {
          if (getAttributeName(attr) !== "set:text") {
            return;
          }
          context.report({
            node: attr.name,
            messageId: "disallow",
            *fix(fixer) {
              const element = attr.parent.parent;
              if (!attr.value || !element || element.type !== "JSXElement") {
                return;
              }
              if (element.children.some(
                (child) => child.type !== "JSXText" || child.value.trim()
              )) {
                return;
              }
              const valueText = attr.type === "AstroTemplateLiteralAttribute" ? `{${sourceCode.getText(attr.value)}}` : sourceCode.getText(attr.value);
              if (element.openingElement.selfClosing) {
                if (sourceCode.text.slice(
                  element.openingElement.range[1] - 2,
                  element.openingElement.range[1]
                ) !== "/>") {
                  return;
                }
                yield fixer.remove(attr);
                yield fixer.removeRange([
                  element.openingElement.range[1] - 2,
                  element.openingElement.range[1] - 1
                ]);
                yield fixer.insertTextAfter(
                  element.openingElement,
                  `${valueText}</${sourceCode.getText(
                    element.openingElement.name
                  )}>`
                );
              } else {
                yield fixer.remove(attr);
                yield* element.children.map((child) => fixer.remove(child));
                yield fixer.insertTextAfter(element.openingElement, valueText);
              }
            }
          });
        }
        return {
          JSXAttribute: verifyName,
          AstroTemplateLiteralAttribute: verifyName
        };
      }
    });
  }
});
function loadModule(context, name2) {
  const sourceCode = getSourceCode(context);
  const key = sourceCode.ast;
  let modules = cache.get(key);
  if (!modules) {
    modules = {};
    cache.set(key, modules);
  }
  const mod = modules[name2];
  if (mod) return mod;
  try {
    const cwd = getCwd(context);
    const relativeTo = path4.join(cwd, "__placeholder__.js");
    return modules[name2] = Module.createRequire(relativeTo)(name2);
  } catch {
    return null;
  }
}
function getContentRange(node) {
  if (node.closingElement) {
    return [node.openingElement.range[1], node.closingElement.range[0]];
  }
  return [node.openingElement.range[1], node.range[1]];
}
var cache;
var init_utils2 = __esm({
  "src/utils/transform/utils.ts"() {
    init_compat();
    cache = /* @__PURE__ */ new WeakMap();
  }
});
function transform(node, context) {
  const postcssLoadConfig = loadPostcssLoadConfig(context);
  if (!postcssLoadConfig) {
    return null;
  }
  const inputRange = getContentRange(node);
  const sourceCode = getSourceCode(context);
  const code = sourceCode.text.slice(...inputRange);
  const filename = `${getFilename(context)}.css`;
  try {
    const config = postcssLoadConfig.sync({
      cwd: getCwd(context) ?? process.cwd(),
      from: filename
    });
    const result = postcss(config.plugins).process(code, {
      ...config.options,
      map: {
        inline: false
      }
    });
    return {
      inputRange,
      output: result.content,
      mappings: result.map.toJSON().mappings
    };
  } catch {
    return null;
  }
}
function loadPostcssLoadConfig(context) {
  return loadModule(context, "postcss-load-config");
}
var init_postcss = __esm({
  "src/utils/transform/postcss.ts"() {
    init_utils2();
    init_compat();
  }
});

// src/utils/transform/sass.ts
function transform2(node, context, type) {
  const sass = loadSass(context);
  if (!sass) {
    return null;
  }
  const inputRange = getContentRange(node);
  const sourceCode = getSourceCode(context);
  const code = sourceCode.text.slice(...inputRange);
  try {
    const output = sass.compileString(code, {
      sourceMap: true,
      syntax: type === "sass" ? "indented" : void 0
    });
    if (!output) {
      return null;
    }
    return {
      inputRange,
      output: output.css,
      mappings: output.sourceMap.mappings
    };
  } catch {
    return null;
  }
}
function loadSass(context) {
  return loadModule(context, "sass");
}
var init_sass = __esm({
  "src/utils/transform/sass.ts"() {
    init_utils2();
    init_compat();
  }
});

// src/utils/transform/less.ts
function transform3(node, context) {
  const less = loadLess(context);
  if (!less) {
    return null;
  }
  const inputRange = getContentRange(node);
  const sourceCode = getSourceCode(context);
  const code = sourceCode.text.slice(...inputRange);
  const filename = `${getFilename(context)}.less`;
  try {
    let output;
    less.render(
      code,
      {
        sourceMap: {},
        syncImport: true,
        filename,
        lint: false
      },
      (_error, result) => {
        output = result;
      }
    );
    if (!output) {
      return null;
    }
    return {
      inputRange,
      output: output.css,
      mappings: JSON.parse(output.map).mappings
    };
  } catch {
    return null;
  }
}
function loadLess(context) {
  return loadModule(context, "less");
}
var init_less = __esm({
  "src/utils/transform/less.ts"() {
    init_utils2();
    init_compat();
  }
});

// src/utils/transform/stylus.ts
function transform4(node, context) {
  const stylus = loadStylus(context);
  if (!stylus) {
    return null;
  }
  const inputRange = getContentRange(node);
  const sourceCode = getSourceCode(context);
  const code = sourceCode.text.slice(...inputRange);
  const filename = `${getFilename(context)}.stylus`;
  try {
    let output;
    const style = stylus(code, {
      filename
    }).set("sourcemap", {});
    style.render((_error, outputCode) => {
      output = outputCode;
    });
    if (output == null) {
      return null;
    }
    return {
      inputRange,
      output,
      mappings: style.sourcemap.mappings
    };
  } catch {
    return null;
  }
}
function loadStylus(context) {
  return loadModule(context, "stylus");
}
var init_stylus = __esm({
  "src/utils/transform/stylus.ts"() {
    init_utils2();
    init_compat();
  }
});

// src/utils/transform/lines-and-columns.ts
function sortedLastIndex(array, value) {
  let lower = 0;
  let upper = array.length;
  while (lower < upper) {
    const mid = Math.floor(lower + (upper - lower) / 2);
    const target = array[mid];
    if (target < value) {
      lower = mid + 1;
    } else if (target > value) {
      upper = mid;
    } else {
      return mid + 1;
    }
  }
  return upper;
}
var LinesAndColumns;
var init_lines_and_columns = __esm({
  "src/utils/transform/lines-and-columns.ts"() {
    LinesAndColumns = class {
      constructor(code) {
        const len = code.length;
        const lineStartIndices = [0];
        for (let index = 0; index < len; index++) {
          const c = code[index];
          if (c === "\r") {
            const next = code[index + 1] || "";
            if (next === "\n") {
              index++;
            }
            lineStartIndices.push(index + 1);
          } else if (c === "\n") {
            lineStartIndices.push(index + 1);
          }
        }
        this.code = code;
        this.lineStartIndices = lineStartIndices;
      }
      getLocFromIndex(index) {
        const lineNumber = sortedLastIndex(this.lineStartIndices, index);
        return {
          line: lineNumber,
          column: index - this.lineStartIndices[lineNumber - 1]
        };
      }
      getIndexFromLoc(loc) {
        const lineIndex = loc.line - 1;
        if (this.lineStartIndices.length > lineIndex) {
          const lineStartIndex = this.lineStartIndices[lineIndex];
          const positionIndex = lineStartIndex + loc.column;
          return positionIndex;
        } else if (this.lineStartIndices.length === lineIndex) {
          return this.code.length + loc.column;
        }
        return this.code.length + loc.column;
      }
    };
  }
});
function getStyleContentCSS(node, context) {
  const cachedResult = cache2.get(node);
  if (cachedResult) {
    return cachedResult;
  }
  const sourceCode = getSourceCode(context);
  const langNode = findAttribute(node, "lang");
  const lang = langNode && getStaticAttributeStringValue(langNode);
  if (!langNode || lang === "css") {
    const inputRange = getContentRange(node);
    return {
      css: sourceCode.text.slice(...inputRange),
      remap: (i) => inputRange[0] + i
    };
  }
  let transform5 = null;
  if (lang === "postcss") {
    transform5 = transform(node, context);
  } else if (lang === "scss" || lang === "sass") {
    transform5 = transform2(node, context, lang);
  } else if (lang === "less") {
    transform5 = transform3(node, context);
  } else if (lang === "styl" || lang === "stylus") {
    transform5 = transform4(node, context);
  }
  if (!transform5) {
    return null;
  }
  const result = transformToStyleContentCSS(transform5, context);
  cache2.set(node, result);
  return result;
}
function transformToStyleContentCSS(transform5, context) {
  const sourceCode = getSourceCode(context);
  let outputLocs = null;
  let inputLocs = null;
  let decoded = null;
  return {
    css: transform5.output,
    remap: (index) => {
      outputLocs = outputLocs ?? new LinesAndColumns(transform5.output);
      inputLocs = inputLocs ?? new LinesAndColumns(sourceCode.text.slice(...transform5.inputRange));
      const outputCodePos = outputLocs.getLocFromIndex(index);
      const inputCodePos = remapPosition(outputCodePos);
      return inputLocs.getIndexFromLoc(inputCodePos) + transform5.inputRange[0];
    }
  };
  function remapPosition(pos) {
    decoded = decoded ?? decode(transform5.mappings);
    const lineMaps = decoded[pos.line - 1];
    if (!lineMaps?.length) {
      for (let line = pos.line - 1; line >= 0; line--) {
        const prevLineMaps = decoded[line];
        if (prevLineMaps?.length) {
          const [, , sourceCodeLine2, sourceCodeColumn2] = prevLineMaps[prevLineMaps.length - 1];
          return {
            line: sourceCodeLine2 + 1,
            column: sourceCodeColumn2
          };
        }
      }
      return { line: -1, column: -1 };
    }
    for (let index = 0; index < lineMaps.length - 1; index++) {
      const [generateCodeColumn2, , sourceCodeLine2, sourceCodeColumn2] = lineMaps[index];
      if (generateCodeColumn2 <= pos.column && pos.column < lineMaps[index + 1][0]) {
        return {
          line: sourceCodeLine2 + 1,
          column: sourceCodeColumn2 + (pos.column - generateCodeColumn2)
        };
      }
    }
    const [generateCodeColumn, , sourceCodeLine, sourceCodeColumn] = lineMaps[lineMaps.length - 1];
    return {
      line: sourceCodeLine + 1,
      column: sourceCodeColumn + (pos.column - generateCodeColumn)
    };
  }
}
var cache2;
var init_transform = __esm({
  "src/utils/transform/index.ts"() {
    init_ast_utils();
    init_utils2();
    init_postcss();
    init_sass();
    init_less();
    init_stylus();
    init_lines_and_columns();
    init_compat();
    cache2 = /* @__PURE__ */ new WeakMap();
  }
});
function parseSelector(selector, context) {
  let astSelector;
  try {
    astSelector = parser().astSync(selector);
  } catch (error) {
    return [
      {
        error,
        selector,
        offset: 0,
        test: () => false
      }
    ];
  }
  return astSelector.nodes.map((sel) => {
    const nodes = removeGlobals(cleanSelectorChildren(sel));
    try {
      const test = selectorToJSXElementMatcher(nodes, context);
      return {
        selector: sel.toString().trim(),
        offset: sel.sourceIndex ?? sel.nodes[0].sourceIndex,
        test(element) {
          return test(element, null);
        }
      };
    } catch (error) {
      if (error instanceof SelectorError) {
        return {
          error,
          selector: sel.toString().trim(),
          offset: sel.sourceIndex ?? sel.nodes[0].sourceIndex,
          test: () => false
        };
      }
      throw error;
    }
  });
  function removeGlobals(nodes) {
    let start = 0;
    let end = nodes.length;
    while (nodes[end - 1] && isGlobalPseudo(nodes[end - 1])) {
      end--;
      if (nodes[end - 1]?.type === "combinator") {
        end--;
      }
    }
    while (nodes[start] && isGlobalPseudo(nodes[start])) {
      start++;
      if (nodes[start]?.type === "combinator") {
        start++;
      }
    }
    if (nodes.some(isRootPseudo)) {
      while (nodes[start] && !isRootPseudo(nodes[start])) {
        start++;
      }
      start++;
      while (nodes[start] && nodes[start].type !== "combinator") {
        start++;
      }
      if (nodes[start]?.type === "combinator") {
        start++;
      }
    }
    return nodes.slice(start, end);
  }
}
function selectorsToJSXElementMatcher(selectorNodes, context) {
  const selectors = selectorNodes.map(
    (n) => selectorToJSXElementMatcher(cleanSelectorChildren(n), context)
  );
  return (element, subject) => selectors.some((sel) => sel(element, subject));
}
function isDescendantCombinator(node) {
  return Boolean(node && node.type === "combinator" && !node.value.trim());
}
function cleanSelectorChildren(selector) {
  const nodes = [];
  let last = null;
  for (const node of selector.nodes) {
    if (node.type === "root") {
      throw new SelectorError("Unexpected state type=root");
    }
    if (node.type === "comment") {
      continue;
    }
    if ((last == null || last.type === "combinator") && isDescendantCombinator(node)) {
      continue;
    }
    if (isDescendantCombinator(last) && node.type === "combinator") {
      nodes.pop();
    }
    nodes.push(node);
    last = node;
  }
  if (isDescendantCombinator(last)) {
    nodes.pop();
  }
  return nodes;
}
function selectorToJSXElementMatcher(selectorChildren, context) {
  const nodes = [...selectorChildren];
  let node = nodes.shift();
  let result = null;
  while (node) {
    if (node.type === "combinator") {
      const combinator = node.value;
      node = nodes.shift();
      if (!node) {
        throw new SelectorError(`Expected selector after '${combinator}'.`);
      }
      if (node.type === "combinator") {
        throw new SelectorError(`Unexpected combinator '${node.value}'.`);
      }
      const right = nodeToJSXElementMatcher(node, context);
      result = combination(
        result || // for :has()
        ((element, subject) => element === subject),
        combinator,
        right
      );
    } else {
      const sel = nodeToJSXElementMatcher(node, context);
      result = result ? compound(result, sel) : sel;
    }
    node = nodes.shift();
  }
  if (!result) {
    return () => true;
  }
  return result;
}
function combination(left, combinator, right) {
  switch (combinator.trim()) {
    case "":
      return (element, subject) => {
        if (right(element, null)) {
          let parent = element.parent;
          while (parent.node) {
            if (left(parent, subject)) {
              return true;
            }
            parent = parent.parent;
          }
        }
        return false;
      };
    case ">":
      return (element, subject) => {
        if (right(element, null)) {
          const parent = element.parent;
          if (parent.node) {
            return left(parent, subject);
          }
        }
        return false;
      };
    case "+":
      return (element, subject) => {
        if (right(element, null)) {
          const before = getBeforeElement(element);
          if (before) {
            return left(before, subject);
          }
        }
        return false;
      };
    case "~":
      return (element, subject) => {
        if (right(element, null)) {
          for (const before of getBeforeElements(element)) {
            if (left(before, subject)) {
              return true;
            }
          }
        }
        return false;
      };
    default:
      throw new SelectorError(`Unknown combinator: ${combinator}.`);
  }
}
function nodeToJSXElementMatcher(selector, context) {
  const baseMatcher = (() => {
    switch (selector.type) {
      case "attribute":
        return attributeNodeToJSXElementMatcher(selector, context);
      case "class":
        return classNameNodeToJSXElementMatcher(selector, context);
      case "id":
        return identifierNodeToJSXElementMatcher(selector, context);
      case "tag":
        return tagNodeToJSXElementMatcher(selector);
      case "universal":
        return universalNodeToJSXElementMatcher();
      case "pseudo":
        return pseudoNodeToJSXElementMatcher(selector, context);
      case "nesting":
        throw new SelectorError("Unsupported nesting selector.");
      case "string":
        throw new SelectorError(`Unknown selector: ${selector.value}.`);
      default:
        throw new SelectorError(`Unknown selector: ${selector.value}.`);
    }
  })();
  return (element, subject) => {
    if (isComponentElement(element)) {
      return false;
    }
    return baseMatcher(element, subject);
  };
}
function attributeNodeToJSXElementMatcher(selector, context) {
  const key = selector.attribute;
  if (!selector.operator) {
    return (element, _) => {
      return hasAttribute(element, key, context);
    };
  }
  const value = selector.value || "";
  switch (selector.operator) {
    case "=":
      return buildJSXElementMatcher(value, (attr, val) => attr === val);
    case "~=":
      return buildJSXElementMatcher(
        value,
        (attr, val) => attr.split(/\s+/u).includes(val)
      );
    case "|=":
      return buildJSXElementMatcher(
        value,
        (attr, val) => attr === val || attr.startsWith(`${val}-`)
      );
    case "^=":
      return buildJSXElementMatcher(value, (attr, val) => attr.startsWith(val));
    case "$=":
      return buildJSXElementMatcher(value, (attr, val) => attr.endsWith(val));
    case "*=":
      return buildJSXElementMatcher(value, (attr, val) => attr.includes(val));
    default:
      throw new SelectorError(`Unsupported operator: ${selector.operator}.`);
  }
  function buildJSXElementMatcher(selectorValue, test) {
    const val = selector.insensitive ? selectorValue.toLowerCase() : selectorValue;
    return (element) => {
      const attr = getAttribute(element, key, context);
      if (attr == null) {
        return false;
      }
      if (attr.unknown || !attr.staticValue) {
        return true;
      }
      const attrValue = attr.staticValue.value;
      return test(
        selector.insensitive ? attrValue.toLowerCase() : attrValue,
        val
      );
    };
  }
}
function classNameNodeToJSXElementMatcher(selector, context) {
  const className = selector.value;
  return (element) => {
    const attr = getAttribute(element, "class", context);
    if (attr == null) {
      return false;
    }
    if (attr.unknown || !attr.staticValue) {
      return true;
    }
    const attrValue = attr.staticValue.value;
    return attrValue.split(/\s+/u).includes(className);
  };
}
function identifierNodeToJSXElementMatcher(selector, context) {
  const id = selector.value;
  return (element) => {
    const attr = getAttribute(element, "id", context);
    if (attr == null) {
      return false;
    }
    if (attr.unknown || !attr.staticValue) {
      return true;
    }
    const attrValue = attr.staticValue.value;
    return attrValue === id;
  };
}
function tagNodeToJSXElementMatcher(selector) {
  const name2 = selector.value;
  return (element) => {
    const elementName = getElementName(element.node);
    return elementName === name2;
  };
}
function universalNodeToJSXElementMatcher(_selector) {
  return () => true;
}
function pseudoNodeToJSXElementMatcher(selector, context) {
  const pseudo = selector.value;
  switch (pseudo) {
    case ":is":
    case ":where":
      return selectorsToJSXElementMatcher(selector.nodes, context);
    case ":has":
      return pseudoHasSelectorsToJSXElementMatcher(selector.nodes, context);
    case ":empty":
      return (element) => element.node.children.every(
        (child) => child.type === "JSXText" && !child.value.trim() || child.type === "AstroHTMLComment"
      );
    // https://docs.astro.build/en/guides/styling/#global-styles
    case ":global": {
      return () => true;
    }
    default:
      return () => true;
  }
}
function pseudoHasSelectorsToJSXElementMatcher(selectorNodes, context) {
  const selectors = selectorNodes.map(
    (n) => pseudoHasSelectorToJSXElementMatcher(n, context)
  );
  return (element, subject) => selectors.some((sel) => sel(element, subject));
}
function pseudoHasSelectorToJSXElementMatcher(selector, context) {
  const nodes = cleanSelectorChildren(selector);
  const selectors = selectorToJSXElementMatcher(nodes, context);
  const firstNode = nodes[0];
  if (firstNode.type === "combinator" && (firstNode.value === "+" || firstNode.value === "~")) {
    return buildJSXElementMatcher((element) => getAfterElements(element));
  }
  return buildJSXElementMatcher((element) => element.childElements);
  function buildJSXElementMatcher(getStartElements) {
    return (element) => {
      const elements = [...getStartElements(element)];
      let curr;
      while (curr = elements.shift()) {
        const el = curr;
        if (selectors(el, element)) {
          return true;
        }
        elements.push(...el.childElements);
      }
      return false;
    };
  }
}
function getBeforeElement(element) {
  return getBeforeElements(element).pop() || null;
}
function getBeforeElements(element) {
  const parent = element.parent;
  if (!parent) {
    return [];
  }
  const index = parent.childElements.indexOf(element);
  return parent.childElements.slice(
    0,
    element.withinExpression ? index + 1 : index
  );
}
function getAfterElements(element) {
  const parent = element.parent;
  if (!parent) {
    return [];
  }
  const index = parent.childElements.indexOf(element);
  return parent.childElements.slice(
    element.withinExpression ? index : index + 1
  );
}
function compound(a, b) {
  return (element, subject) => a(element, subject) && b(element, subject);
}
function isComponentElement(element) {
  const elementName = getElementName(element.node);
  return elementName == null || elementName.toLowerCase() !== elementName;
}
function isGlobalPseudo(node) {
  return node.type === "pseudo" && node.value === ":global";
}
function isRootPseudo(node) {
  return node.type === "pseudo" && node.value === ":root";
}
function hasAttribute(element, attribute, context) {
  const attr = getAttribute(element, attribute, context);
  if (attr) {
    return true;
  }
  return false;
}
function getAttribute(element, attribute, context) {
  const attr = findAttribute(element.node, attribute);
  if (attr) {
    if (attr.value == null) {
      return {
        unknown: false,
        hasAttr: true,
        staticValue: { value: "" }
      };
    }
    const value = getStaticAttributeStringValue(attr, context);
    if (value == null) {
      return {
        unknown: false,
        hasAttr: true,
        staticValue: null
      };
    }
    return {
      unknown: false,
      hasAttr: true,
      staticValue: { value }
    };
  }
  if (attribute === "class") {
    const result = getClassListAttribute(element, context);
    if (result) {
      return result;
    }
  }
  const spreadAttributes = getSpreadAttributes(element.node);
  if (spreadAttributes.length === 0) {
    return null;
  }
  return {
    unknown: true
  };
}
function getClassListAttribute(element, context) {
  const attr = findAttribute(element.node, "class:list");
  if (attr) {
    if (attr.value == null) {
      return {
        unknown: false,
        hasAttr: true,
        staticValue: { value: "" }
      };
    }
    const classList = extractClassList(attr, context);
    if (classList === null) {
      return {
        unknown: false,
        hasAttr: true,
        staticValue: null
      };
    }
    return {
      unknown: false,
      hasAttr: true,
      staticValue: { value: classList.classList.join(" ") }
    };
  }
  return null;
}
function extractClassList(node, context) {
  if (node.value?.type === AST_NODE_TYPES.Literal) {
    return { classList: [String(node.value.value)] };
  }
  if (node.value?.type === "JSXExpressionContainer" && node.value.expression.type !== "JSXEmptyExpression") {
    const classList = [];
    for (const className of extractClassListFromExpression(
      node.value.expression,
      context
    )) {
      if (className == null) {
        return null;
      }
      classList.push(className);
    }
    return { classList };
  }
  return null;
}
function* extractClassListFromExpression(node, context) {
  if (node.type === AST_NODE_TYPES.ArrayExpression) {
    for (const element of node.elements) {
      if (element == null) continue;
      if (element.type === AST_NODE_TYPES.SpreadElement) {
        yield* extractClassListFromExpression(element.argument, context);
      } else {
        yield* extractClassListFromExpression(element, context);
      }
    }
    return;
  }
  if (node.type === AST_NODE_TYPES.ObjectExpression) {
    for (const prop of node.properties) {
      if (prop.type === AST_NODE_TYPES.SpreadElement) {
        yield* extractClassListFromExpression(prop.argument, context);
      } else if (!prop.computed) {
        if (prop.key.type === AST_NODE_TYPES.Literal) {
          yield String(prop.key.value);
        } else {
          yield prop.key.name;
        }
      } else {
        yield* extractClassListFromExpression(prop.key, context);
      }
    }
    return;
  }
  const sourceCode = getSourceCode(context);
  const staticValue = getStaticValue(
    node,
    sourceCode.scopeManager.globalScope
  );
  if (staticValue) {
    yield* extractClassListFromUnknown(staticValue.value);
    return;
  }
  yield null;
}
function* extractClassListFromUnknown(value) {
  if (!value) {
    return;
  }
  if (Array.isArray(value)) {
    for (const e of value) {
      yield* extractClassListFromUnknown(e);
    }
    return;
  }
  if (typeof value === "object") {
    yield* Object.keys(value);
    return;
  }
  yield String(value);
}
var no_unused_css_selector_default, SelectorError;
var init_no_unused_css_selector = __esm({
  "src/rules/no-unused-css-selector.ts"() {
    init_utils();
    init_ast_utils();
    init_transform();
    init_compat();
    no_unused_css_selector_default = createRule("no-unused-css-selector", {
      meta: {
        docs: {
          description: "disallow selectors defined in `style` tag that don't use in HTML",
          category: "Best Practices",
          recommended: false
        },
        schema: [],
        messages: {
          unused: "Unused CSS selector `{{selector}}`"
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        const styles = [];
        const rootTree = {
          parent: null,
          node: null,
          childElements: []
        };
        const allTreeElements = [];
        let currTree = rootTree;
        function verifyCSS(css) {
          let root;
          try {
            root = postcss.parse(css.css);
          } catch {
            return;
          }
          const ignoreNodes = /* @__PURE__ */ new Set();
          root.walk((psNode) => {
            if (psNode.parent && ignoreNodes.has(psNode.parent)) {
              ignoreNodes.add(psNode);
              return;
            }
            if (psNode.type !== "rule") {
              if (psNode.type === "atrule") {
                if (psNode.name === "keyframes") {
                  ignoreNodes.add(psNode);
                }
              }
              return;
            }
            const rule = psNode;
            const raws = rule.raws;
            const rawSelectorText = raws.selector ? raws.selector.raw : rule.selector;
            for (const selector of parseSelector(rawSelectorText, context)) {
              if (selector.error) {
                continue;
              }
              if (allTreeElements.some((tree) => selector.test(tree))) {
                continue;
              }
              reportSelector(
                rule.source.start.offset + selector.offset,
                selector.selector
              );
            }
          });
          function reportSelector(start, selector) {
            const remapStart = css.remap(start);
            const remapEnd = css.remap(start + selector.length);
            context.report({
              loc: {
                start: sourceCode.getLocFromIndex(remapStart),
                end: sourceCode.getLocFromIndex(remapEnd)
              },
              messageId: "unused",
              data: {
                selector
              }
            });
          }
        }
        return {
          JSXElement(node) {
            const name2 = getElementName(node);
            if (name2 === "Fragment" || name2 === "slot") {
              return;
            }
            if (name2 === "style" && !findAttribute(node, "is:global")) {
              styles.push(node);
            }
            const tree = {
              parent: currTree,
              node,
              childElements: []
            };
            allTreeElements.unshift(tree);
            currTree.childElements.push(tree);
            currTree = tree;
          },
          "JSXElement:exit"(node) {
            if (currTree.node === node) {
              if (currTree.node) {
                const expressions = currTree.node.children.filter(
                  (e) => e.type === "JSXExpressionContainer"
                );
                if (expressions.length) {
                  for (const child of currTree.childElements) {
                    child.withinExpression = expressions.some(
                      (e) => e.range[0] <= child.node.range[0] && child.node.range[1] <= e.range[1]
                    );
                  }
                }
              }
              currTree = currTree.parent;
            }
          },
          "Program:exit"() {
            for (const style of styles) {
              const css = getStyleContentCSS(style, context);
              if (css) {
                verifyCSS(css);
              }
            }
          }
        };
      }
    });
    SelectorError = class extends Error {
    };
  }
});

// src/utils/style/tokenizer.ts
function isWhitespace(cp) {
  return cp === TABULATION || cp === LINE_FEED || cp === FORM_FEED || cp === CARRIAGE_RETURN || cp === SPACE;
}
function isPunctuator(cp) {
  return cp === COLON || cp === SEMICOLON || cp === COMMA || // Brackets
  cp === LEFT_PARENTHESIS || cp === RIGHT_PARENTHESIS || cp === LEFT_CURLY_BRACKET || cp === RIGHT_CURLY_BRACKET || cp === LEFT_SQUARE_BRACKET || cp === RIGHT_SQUARE_BRACKET || // Maybe v-bind() in calc()
  cp === SOLIDUS || cp === ASTERISK;
}
function isQuote(cp) {
  return cp === APOSTROPHE || cp === QUOTATION_MARK;
}
var EOF, NULL, TABULATION, CARRIAGE_RETURN, LINE_FEED, FORM_FEED, SPACE, QUOTATION_MARK, APOSTROPHE, LEFT_PARENTHESIS, RIGHT_PARENTHESIS, ASTERISK, COMMA, SOLIDUS, COLON, SEMICOLON, LEFT_SQUARE_BRACKET, REVERSE_SOLIDUS, RIGHT_SQUARE_BRACKET, LEFT_CURLY_BRACKET, RIGHT_CURLY_BRACKET, CSSTokenizer;
var init_tokenizer = __esm({
  "src/utils/style/tokenizer.ts"() {
    EOF = -1;
    NULL = 0;
    TABULATION = 9;
    CARRIAGE_RETURN = 13;
    LINE_FEED = 10;
    FORM_FEED = 12;
    SPACE = 32;
    QUOTATION_MARK = 34;
    APOSTROPHE = 39;
    LEFT_PARENTHESIS = 40;
    RIGHT_PARENTHESIS = 41;
    ASTERISK = 42;
    COMMA = 44;
    SOLIDUS = 47;
    COLON = 58;
    SEMICOLON = 59;
    LEFT_SQUARE_BRACKET = 91;
    REVERSE_SOLIDUS = 92;
    RIGHT_SQUARE_BRACKET = 93;
    LEFT_CURLY_BRACKET = 123;
    RIGHT_CURLY_BRACKET = 125;
    CSSTokenizer = class {
      /**
       * Initialize this tokenizer.
       * @param text The source code to tokenize.
       * @param options The tokenizer options.
       */
      constructor(text, startOffset, options) {
        this.text = text;
        this.options = {
          inlineComment: options?.inlineComment ?? false
        };
        this.cp = NULL;
        this.offset = startOffset - 1;
        this.nextOffset = startOffset;
        this.reconsuming = false;
      }
      /**
       * Get the next token.
       * @returns The next token or null.
       */
      nextToken() {
        let cp;
        if (this.reconsuming) {
          cp = this.cp;
          this.reconsuming = false;
        } else {
          cp = this.consumeNextCodePoint();
        }
        while (isWhitespace(cp)) {
          cp = this.consumeNextCodePoint();
        }
        if (cp === EOF) {
          return null;
        }
        const start = this.offset;
        return this.consumeNextToken(cp, start);
      }
      /**
       * Get the next code point.
       * @returns The code point.
       */
      nextCodePoint() {
        if (this.nextOffset >= this.text.length) {
          return EOF;
        }
        return this.text.codePointAt(this.nextOffset);
      }
      /**
       * Consume the next code point.
       * @returns The consumed code point.
       */
      consumeNextCodePoint() {
        if (this.offset >= this.text.length) {
          this.cp = EOF;
          return EOF;
        }
        this.offset = this.nextOffset;
        if (this.offset >= this.text.length) {
          this.cp = EOF;
          return EOF;
        }
        let cp = this.text.codePointAt(this.offset);
        if (cp === CARRIAGE_RETURN) {
          this.nextOffset = this.offset + 1;
          if (this.text.codePointAt(this.nextOffset) === LINE_FEED) {
            this.nextOffset++;
          }
          cp = LINE_FEED;
        } else {
          this.nextOffset = this.offset + (cp >= 65536 ? 2 : 1);
        }
        this.cp = cp;
        return cp;
      }
      consumeNextToken(cp, start) {
        if (cp === SOLIDUS) {
          const nextCp = this.nextCodePoint();
          if (nextCp === ASTERISK) {
            return this.consumeComment(start);
          }
          if (nextCp === SOLIDUS && this.options.inlineComment) {
            return this.consumeInlineComment(start);
          }
        }
        if (isQuote(cp)) {
          return this.consumeString(start, cp);
        }
        if (isPunctuator(cp)) {
          return {
            type: "Punctuator" /* punctuator */,
            range: [start, start + 1],
            value: String.fromCodePoint(cp)
          };
        }
        return this.consumeWord(start);
      }
      /**
       * Consume word
       */
      consumeWord(start) {
        let cp = this.consumeNextCodePoint();
        while (!isWhitespace(cp) && !isPunctuator(cp) && !isQuote(cp)) {
          cp = this.consumeNextCodePoint();
        }
        this.reconsuming = true;
        const range = [start, this.offset];
        const text = this.text;
        let value;
        return {
          type: "Word" /* word */,
          range,
          get value() {
            return value ?? (value = text.slice(...range));
          }
        };
      }
      /**
       * https://drafts.csswg.org/css-syntax/#consume-string-token
       */
      consumeString(start, quote) {
        let valueEndOffset = null;
        let cp = this.consumeNextCodePoint();
        while (cp !== EOF) {
          if (cp === quote) {
            valueEndOffset = this.offset;
            break;
          }
          if (cp === REVERSE_SOLIDUS) {
            this.consumeNextCodePoint();
          }
          cp = this.consumeNextCodePoint();
        }
        const text = this.text;
        let value;
        const valueRange = [
          start + 1,
          valueEndOffset ?? this.nextOffset
        ];
        return {
          type: "Quoted" /* quoted */,
          range: [start, this.nextOffset],
          valueRange,
          get value() {
            return value ?? (value = text.slice(...valueRange));
          },
          quote: String.fromCodePoint(quote)
        };
      }
      /**
       * https://drafts.csswg.org/css-syntax/#consume-comment
       */
      consumeComment(start) {
        this.consumeNextCodePoint();
        let valueEndOffset = null;
        let cp = this.consumeNextCodePoint();
        while (cp !== EOF) {
          if (cp === ASTERISK) {
            cp = this.consumeNextCodePoint();
            if (cp === SOLIDUS) {
              valueEndOffset = this.offset - 1;
              break;
            }
          }
          cp = this.consumeNextCodePoint();
        }
        const valueRange = [
          start + 2,
          valueEndOffset ?? this.nextOffset
        ];
        const text = this.text;
        let value;
        return {
          type: "Block" /* block */,
          range: [start, this.nextOffset],
          valueRange,
          get value() {
            return value ?? (value = text.slice(...valueRange));
          }
        };
      }
      /**
       * Consume inline comment
       */
      consumeInlineComment(start) {
        this.consumeNextCodePoint();
        let valueEndOffset = null;
        let cp = this.consumeNextCodePoint();
        while (cp !== EOF) {
          if (cp === LINE_FEED) {
            valueEndOffset = this.offset - 1;
            break;
          }
          cp = this.consumeNextCodePoint();
        }
        const valueRange = [
          start + 2,
          valueEndOffset ?? this.nextOffset
        ];
        const text = this.text;
        let value;
        return {
          type: "Line" /* line */,
          range: [start, this.nextOffset],
          valueRange,
          get value() {
            return value ?? (value = text.slice(...valueRange));
          }
        };
      }
    };
  }
});

// src/utils/style/index.ts
function* iterateCSSVars(code, cssOptions) {
  const tokenizer = new CSSTokenScanner(code, cssOptions);
  let token;
  while (token = tokenizer.nextToken()) {
    if (token.type === "Word" /* word */ || token.value.startsWith("--")) {
      yield token.value;
    }
  }
}
var CSSTokenScanner;
var init_style = __esm({
  "src/utils/style/index.ts"() {
    init_tokenizer();
    CSSTokenScanner = class {
      constructor(text, options) {
        this.reconsuming = [];
        this.tokenizer = new CSSTokenizer(text, 0, options);
      }
      nextToken() {
        return this.reconsuming.shift() || this.tokenizer.nextToken();
      }
      reconsume(...tokens) {
        this.reconsuming.push(...tokens);
      }
    };
  }
});
var no_unused_define_vars_in_style_default;
var init_no_unused_define_vars_in_style = __esm({
  "src/rules/no-unused-define-vars-in-style.ts"() {
    init_utils();
    init_ast_utils();
    init_style();
    init_compat();
    no_unused_define_vars_in_style_default = createRule("no-unused-define-vars-in-style", {
      meta: {
        docs: {
          description: "disallow unused `define:vars={...}` in `style` tag",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {
          unused: "'{{varName}}' is defined but never used."
        },
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          "JSXElement > JSXOpeningElement[name.type='JSXIdentifier'][name.name='style']"(node) {
            const defineVars = node.attributes.find(
              (attr) => getAttributeName(attr) === "define:vars"
            );
            if (!defineVars) {
              return;
            }
            if (!defineVars.value || defineVars.value.type !== AST_NODE_TYPES.JSXExpressionContainer || defineVars.value.expression.type !== AST_NODE_TYPES.ObjectExpression) {
              return;
            }
            if (node.parent.children.length !== 1) {
              return;
            }
            const textNode = node.parent.children[0];
            if (!textNode || textNode.type !== "AstroRawText") {
              return;
            }
            const definedVars = defineVars.value.expression.properties.filter(
              (prop) => prop.type === AST_NODE_TYPES.Property
            ).map((prop) => ({
              prop,
              name: getPropertyName(prop, sourceCode.getScope(node))
            })).filter(
              (data) => Boolean(data.name)
            );
            if (!definedVars.length) {
              return;
            }
            const lang = node.attributes.find(
              (attr) => getAttributeName(attr) === "lang"
            );
            const langValue = lang && lang.value && lang.value.type === AST_NODE_TYPES.Literal && lang.value.value;
            let unusedDefinedVars = [...definedVars];
            for (const cssVar of iterateCSSVars(textNode.value, {
              inlineComment: Boolean(langValue) && langValue !== "css"
            })) {
              const variable = cssVar.slice(2);
              unusedDefinedVars = unusedDefinedVars.filter(
                (v) => v.name !== variable
              );
            }
            for (const unused of unusedDefinedVars) {
              context.report({
                node: unused.prop.key,
                messageId: "unused",
                data: {
                  varName: unused.name
                }
              });
            }
          }
        };
      }
    });
  }
});

// src/rules/prefer-class-list-directive.ts
var prefer_class_list_directive_default;
var init_prefer_class_list_directive = __esm({
  "src/rules/prefer-class-list-directive.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    prefer_class_list_directive_default = createRule("prefer-class-list-directive", {
      meta: {
        docs: {
          description: "require `class:list` directives instead of `class` with expressions",
          category: "Stylistic Issues",
          recommended: false
        },
        schema: [],
        messages: {
          unexpected: "Unexpected `class` using expression. Use 'class:list' instead."
        },
        fixable: "code",
        type: "suggestion"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        function verifyAttr(attr) {
          if (getAttributeName(attr) !== "class") {
            return;
          }
          if (!attr.value || attr.value.type !== "JSXExpressionContainer" || attr.value.expression.type === "JSXEmptyExpression") {
            return;
          }
          context.report({
            node: attr.name,
            messageId: "unexpected",
            fix(fixer) {
              if (attr.type === "AstroShorthandAttribute") {
                return fixer.insertTextBefore(attr, "class:list=");
              }
              return fixer.insertTextAfter(attr.name, ":list");
            }
          });
        }
        return {
          JSXAttribute: verifyAttr,
          AstroTemplateLiteralAttribute: verifyAttr
        };
      }
    });
  }
});
var prefer_object_class_list_default;
var init_prefer_object_class_list = __esm({
  "src/rules/prefer-object-class-list.ts"() {
    init_utils();
    init_ast_utils();
    init_compat();
    prefer_object_class_list_default = createRule("prefer-object-class-list", {
      meta: {
        docs: {
          description: "require use object instead of ternary expression in `class:list`",
          category: "Stylistic Issues",
          recommended: false
        },
        schema: [],
        messages: {
          unexpected: "Unexpected class using the ternary operator."
        },
        fixable: "code",
        type: "suggestion"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        class NewObjectProps {
          constructor() {
            this.props = [];
          }
          toObjectString() {
            return `{${this.toPropsString()}}`;
          }
          fixObject({
            fixer,
            object
          }) {
            const closeBrace = sourceCode.getLastToken(object);
            const maybeComma = sourceCode.getTokenBefore(closeBrace);
            let text;
            if (isCommaToken(maybeComma)) {
              text = this.toPropsString();
            } else {
              text = `,${this.toPropsString()}`;
            }
            return fixer.insertTextAfterRange(maybeComma.range, text);
          }
          toPropsString() {
            return `${this.props.map(({ key, value }) => `${key}: ${value}`).join(", ")}`;
          }
        }
        function parseConditionalExpression(node) {
          const result = /* @__PURE__ */ new Map();
          if (!processItems(
            {
              node: node.test
            },
            node.consequent
          )) {
            return null;
          }
          if (!processItems(
            {
              not: true,
              node: node.test
            },
            node.alternate
          )) {
            return null;
          }
          return result;
          function processItems(key, e) {
            if (e.type === "ConditionalExpression") {
              const sub = parseConditionalExpression(e);
              if (sub == null) {
                return false;
              }
              for (const [expr, str] of sub) {
                result.set(
                  {
                    ...key,
                    chains: expr
                  },
                  str
                );
              }
            } else {
              const str = getStringIfConstant(e);
              if (str == null) {
                return false;
              }
              result.set(key, str);
            }
            return true;
          }
        }
        function exprToString({ node, not }) {
          let text = sourceCode.text.slice(...node.range);
          if (not) {
            if (node.type === "BinaryExpression") {
              if (node.operator === "===" || node.operator === "==" || node.operator === "!==" || node.operator === "!=") {
                const left = sourceCode.text.slice(...node.left.range);
                const op = sourceCode.text.slice(
                  node.left.range[1],
                  node.right.range[0]
                );
                const right = sourceCode.text.slice(...node.right.range);
                return `${left}${node.operator === "===" || node.operator === "==" ? op.replace(/[=](={1,2})/g, "!$1") : op.replace(/!(={1,2})/g, "=$1")}${right}`;
              }
            } else if (node.type === "UnaryExpression") {
              if (node.operator === "!" && node.prefix) {
                return sourceCode.text.slice(...node.argument.range);
              }
            }
            if (needParentheses(node)) {
              text = `(${text})`;
            }
            text = `!${text}`;
          }
          return text;
        }
        function getStrings(node) {
          if (node.type === "TemplateElement") {
            return [node.value.cooked];
          }
          if (node.type === "ConditionalExpression") {
            const values = parseConditionalExpression(node);
            if (values == null) {
              return null;
            }
            return [...values.values()];
          }
          const str = getStringIfConstant(node);
          if (str == null) {
            return null;
          }
          return [str];
        }
        function endsWithSpace(elements) {
          for (let i = elements.length - 1; i >= 0; i--) {
            const valueNode = elements[i];
            const strings = getStrings(valueNode);
            if (strings == null) {
              if (valueNode.type === AST_NODE_TYPES.TemplateLiteral) {
                const quasiValue = valueNode.quasis[valueNode.quasis.length - 1].value.cooked;
                if (quasiValue && !quasiValue[quasiValue.length - 1].trim()) {
                  return true;
                }
              }
              return false;
            }
            let hasEmpty = false;
            for (const str of strings) {
              if (str) {
                if (str[str.length - 1].trim()) {
                  return false;
                }
              } else {
                hasEmpty = true;
              }
            }
            if (!hasEmpty) {
              return true;
            }
          }
          return null;
        }
        function startsWithSpace(elements) {
          for (let i = 0; i < elements.length; i++) {
            const valueNode = elements[i];
            const strings = getStrings(valueNode);
            if (strings == null) {
              if (valueNode.type === AST_NODE_TYPES.TemplateLiteral) {
                const quasiValue = valueNode.quasis[0].value.cooked;
                if (quasiValue && !quasiValue[0].trim()) {
                  return true;
                }
              }
              return false;
            }
            let hasEmpty = false;
            for (const str of strings) {
              if (str) {
                if (str[0].trim()) {
                  return false;
                }
              } else {
                hasEmpty = true;
              }
            }
            if (!hasEmpty) {
              return true;
            }
          }
          return null;
        }
        function report(node, map, state) {
          context.report({
            node,
            messageId: "unexpected",
            *fix(fixer) {
              const classProps = new NewObjectProps();
              let beforeSpaces = "";
              let afterSpaces = "";
              for (const [expr, className] of map) {
                const trimmedClassName = className.trim();
                if (trimmedClassName) {
                  classProps.props.push({
                    key: JSON.stringify(trimmedClassName),
                    value: exprToString(expr)
                  });
                } else if (!classProps.props.length) {
                  beforeSpaces += className;
                } else {
                  afterSpaces += className;
                }
              }
              yield* state.fixExpression({
                newProps: classProps,
                beforeSpaces,
                afterSpaces,
                node,
                fixer
              });
            }
          });
        }
        function verifyConditionalExpression(node, state) {
          const map = parseConditionalExpression(node);
          if (map == null) {
            return;
          }
          let canTransform = true;
          for (const className of map.values()) {
            if (className) {
              if (className[0].trim() && state.beforeIsWord() || className[className.length - 1].trim() && state.afterIsWord()) {
                canTransform = false;
                break;
              }
            } else if (state.beforeIsWord() && state.afterIsWord()) {
              canTransform = false;
              break;
            }
          }
          if (!canTransform) {
            return;
          }
          report(node, map, state);
        }
        function verifyAttr(attr) {
          if (getAttributeName(attr) !== "class:list") {
            return;
          }
          if (!attr.value || attr.value.type !== AST_NODE_TYPES.JSXExpressionContainer || attr.value.expression.type === AST_NODE_TYPES.JSXEmptyExpression) {
            return;
          }
          const expression = attr.value.expression;
          for (const element of extractElements(expression)) {
            visitElementExpression(element.node, {
              beforeIsWord: () => false,
              afterIsWord: () => false,
              *fixArrayElement(data) {
                yield data.fixer.removeRange(
                  getParenthesizedRange(data.node, sourceCode)
                );
                if (!element.array) {
                  let open, close;
                  if (attr.type === "AstroTemplateLiteralAttribute") {
                    open = "{[";
                    close = "]}";
                  } else {
                    open = "[";
                    close = "]";
                  }
                  yield data.fixer.insertTextBeforeRange(expression.range, open);
                  yield data.fixer.insertTextAfterRange(
                    expression.range,
                    `,${data.newProps.toObjectString()}${close}`
                  );
                  return;
                }
                const object = findClosestObject(element.array, element.node);
                if (object) {
                  yield data.newProps.fixObject({ fixer: data.fixer, object });
                  return;
                }
                const tokens = getParenthesizedTokens(element.node, sourceCode);
                const maybeComma = sourceCode.getTokenAfter(tokens.right);
                let insertOffset, text;
                if (isCommaToken(maybeComma)) {
                  insertOffset = maybeComma.range[1];
                  text = data.newProps.toObjectString();
                } else {
                  insertOffset = tokens.right.range[1];
                  text = `,${data.newProps.toObjectString()}`;
                }
                if (element.array.elements[element.array.elements.length - 1] !== element.node) {
                  text += ",";
                }
                yield data.fixer.insertTextAfterRange(
                  [insertOffset, insertOffset],
                  text
                );
              },
              *fixExpression(data) {
                if (element.array) {
                  const object = findClosestObject(element.array, element.node);
                  if (object) {
                    yield data.fixer.removeRange(
                      getParenthesizedRange(data.node, sourceCode)
                    );
                    const tokens = getParenthesizedTokens(element.node, sourceCode);
                    const maybeComma = sourceCode.getTokenAfter(tokens.right);
                    if (isCommaToken(maybeComma)) {
                      yield data.fixer.removeRange(maybeComma.range);
                    } else {
                      const maybeBeforeComma = sourceCode.getTokenBefore(
                        tokens.left
                      );
                      if (isCommaToken(maybeBeforeComma)) {
                        yield data.fixer.removeRange(maybeBeforeComma.range);
                      }
                    }
                    yield data.newProps.fixObject({ fixer: data.fixer, object });
                    return;
                  }
                }
                yield data.fixer.replaceTextRange(
                  getParenthesizedRange(data.node, sourceCode),
                  data.newProps.toObjectString()
                );
              }
            });
          }
          function findClosestObject(array, target) {
            const index = array.elements.indexOf(target);
            const afterElements = array.elements.slice(index + 1);
            const beforeElements = array.elements.slice(0, index).reverse();
            const length = Math.max(afterElements.length, beforeElements.length);
            for (let index2 = 0; index2 < length; index2++) {
              const after = afterElements[index2];
              if (after?.type === AST_NODE_TYPES.ObjectExpression) {
                return after;
              }
              const before = beforeElements[index2];
              if (before?.type === AST_NODE_TYPES.ObjectExpression) {
                return before;
              }
            }
            return null;
          }
          function visitElementExpression(node, state) {
            if (node.type === AST_NODE_TYPES.ConditionalExpression) {
              verifyConditionalExpression(node, state);
            } else if (node.type === AST_NODE_TYPES.TemplateLiteral) {
              const quasis = [...node.quasis];
              let beforeQuasiWk = quasis.shift();
              for (const expression2 of node.expressions) {
                const beforeQuasi = beforeQuasiWk;
                const afterQuasi = quasis.shift();
                visitElementExpression(expression2, {
                  beforeIsWord() {
                    const beforeElements = [];
                    const targetIndex = node.expressions.indexOf(expression2);
                    for (let index = 0; index < targetIndex; index++) {
                      beforeElements.push(
                        node.quasis[index],
                        node.expressions[index]
                      );
                    }
                    beforeElements.push(node.quasis[targetIndex]);
                    const isSpace = endsWithSpace(beforeElements);
                    return isSpace == null ? state.beforeIsWord() : !isSpace;
                  },
                  afterIsWord() {
                    const targetIndex = node.expressions.indexOf(expression2);
                    const afterElements = [node.quasis[targetIndex + 1]];
                    for (let index = targetIndex + 1; index < node.expressions.length; index++) {
                      afterElements.push(
                        node.expressions[index],
                        node.quasis[index + 1]
                      );
                    }
                    const isSpace = startsWithSpace(afterElements);
                    return isSpace == null ? state.afterIsWord() : !isSpace;
                  },
                  fixArrayElement: state.fixArrayElement,
                  *fixExpression(data) {
                    const fixer = data.fixer;
                    if (beforeQuasi.value.cooked.trim() || afterQuasi.value.cooked.trim() || // has other expression
                    node.expressions.length > 1) {
                      yield fixer.replaceTextRange(
                        [beforeQuasi.range[1] - 2, beforeQuasi.range[1]],
                        data.beforeSpaces
                      );
                      yield fixer.replaceTextRange(
                        [afterQuasi.range[0], afterQuasi.range[0] + 1],
                        data.afterSpaces
                      );
                      yield* state.fixArrayElement(data);
                      return;
                    }
                    const tokens = getParenthesizedTokens(node, sourceCode);
                    yield fixer.removeRange([
                      tokens.left.range[0],
                      beforeQuasi.range[1]
                    ]);
                    yield fixer.removeRange([
                      afterQuasi.range[0],
                      tokens.right.range[1]
                    ]);
                    yield* state.fixExpression({
                      ...data,
                      beforeSpaces: beforeQuasi.value.cooked + data.beforeSpaces,
                      afterSpaces: data.afterSpaces + afterQuasi.value.cooked
                    });
                  }
                });
                beforeQuasiWk = afterQuasi;
              }
            } else if (node.type === AST_NODE_TYPES.CallExpression) {
              if (isStringCallExpression(node) && node.arguments[0] && node.arguments[0].type !== AST_NODE_TYPES.SpreadElement) {
                visitElementExpression(node.arguments[0], {
                  beforeIsWord: state.beforeIsWord,
                  afterIsWord: state.afterIsWord,
                  fixArrayElement: state.fixArrayElement,
                  *fixExpression(data) {
                    const openParen = sourceCode.getTokenAfter(
                      getParenthesizedTokens(node.callee, sourceCode).right
                    );
                    const stripStart = sourceCode.getTokenAfter(
                      getParenthesizedTokens(node.arguments[0], sourceCode).right
                    );
                    const tokens = getParenthesizedTokens(node, sourceCode);
                    yield data.fixer.removeRange([
                      tokens.left.range[0],
                      openParen.range[1]
                    ]);
                    yield data.fixer.removeRange([
                      stripStart.range[0],
                      tokens.right.range[1]
                    ]);
                    yield* state.fixExpression(data);
                  }
                });
              } else if (node.callee.type === AST_NODE_TYPES.MemberExpression && getPropertyName(node.callee) === "trim") {
                const men = node.callee;
                visitElementExpression(men.object, {
                  beforeIsWord: state.beforeIsWord,
                  afterIsWord: state.afterIsWord,
                  fixArrayElement: state.fixArrayElement,
                  *fixExpression(data) {
                    const tokens = getParenthesizedTokens(men.object, sourceCode);
                    yield data.fixer.removeRange([
                      tokens.right.range[1],
                      node.range[1]
                    ]);
                    yield* state.fixExpression(data);
                  }
                });
              }
            } else if (node.type === AST_NODE_TYPES.BinaryExpression) {
              const elements = extractConcatExpressions(node, sourceCode);
              if (!elements) {
                return;
              }
              for (const expression2 of elements) {
                visitElementExpression(expression2, {
                  beforeIsWord() {
                    const index = elements.indexOf(expression2);
                    const beforeElements = elements.slice(0, index);
                    const isSpace = endsWithSpace(beforeElements);
                    return isSpace == null ? state.beforeIsWord() : !isSpace;
                  },
                  afterIsWord() {
                    const index = elements.indexOf(expression2);
                    const afterElements = elements.slice(index + 1);
                    const isSpace = startsWithSpace(afterElements);
                    return isSpace == null ? state.afterIsWord() : !isSpace;
                  },
                  fixArrayElement: state.fixArrayElement,
                  *fixExpression(data) {
                    const fixer = data.fixer;
                    const index = elements.indexOf(expression2);
                    const beforeElements = elements.slice(0, index);
                    const afterElements = elements.slice(index + 1);
                    const tokens = getParenthesizedTokens(expression2, sourceCode);
                    if (beforeElements.some((element) => {
                      const str = getStringIfConstant(element);
                      return str == null || Boolean(str.trim());
                    }) || afterElements.some((element) => {
                      const str = getStringIfConstant(element);
                      return str == null || Boolean(str.trim());
                    })) {
                      const beforeElement = beforeElements[beforeElements.length - 1];
                      const afterElement = afterElements[0];
                      if (beforeElement && isStringLiteral(beforeElement) && afterElement && isStringLiteral(afterElement)) {
                        if (sourceCode.text[beforeElement.range[0]] !== sourceCode.text[afterElement.range[0]]) {
                          const targetIsBefore = sourceCode.text[beforeElement.range[0]] === "'";
                          const replaceLiteral = targetIsBefore ? beforeElement : afterElement;
                          yield fixer.replaceTextRange(
                            [
                              replaceLiteral.range[0] + 1,
                              replaceLiteral.range[1] - 1
                            ],
                            JSON.stringify(replaceLiteral.value).slice(1, -1)
                          );
                          yield fixer.replaceTextRange(
                            targetIsBefore ? [
                              replaceLiteral.range[0],
                              replaceLiteral.range[0] + 1
                            ] : [
                              replaceLiteral.range[1] - 1,
                              replaceLiteral.range[1]
                            ],
                            '"'
                          );
                        }
                        yield fixer.replaceTextRange(
                          [beforeElement.range[1] - 1, tokens.left.range[0]],
                          data.beforeSpaces
                        );
                        yield fixer.replaceTextRange(
                          [tokens.right.range[1], afterElement.range[0] + 1],
                          data.afterSpaces
                        );
                      } else {
                        const beforeToken = sourceCode.getTokenBefore(tokens.left);
                        if (beforeToken?.value === "+") {
                          yield fixer.removeRange(beforeToken.range);
                        } else {
                          const afterToken = sourceCode.getTokenAfter(tokens.right);
                          yield fixer.removeRange(afterToken.range);
                        }
                      }
                      yield* state.fixArrayElement(data);
                      return;
                    }
                    if (beforeElements.length) {
                      const beforeToken = sourceCode.getTokenBefore(tokens.left);
                      yield fixer.removeRange([
                        beforeElements[0].range[0],
                        beforeToken.range[1]
                      ]);
                    }
                    if (afterElements.length) {
                      const afterToken = sourceCode.getTokenAfter(tokens.right);
                      yield fixer.removeRange([
                        afterToken.range[0],
                        afterElements[afterElements.length - 1].range[1]
                      ]);
                    }
                    yield* state.fixExpression({
                      ...data,
                      beforeSpaces: beforeElements.map((e) => getStringIfConstant(e)).join("") + data.beforeSpaces,
                      afterSpaces: data.afterSpaces + afterElements.map((e) => getStringIfConstant(e)).join("")
                    });
                  }
                });
              }
            }
          }
        }
        function extractElements(node) {
          if (node.type === AST_NODE_TYPES.ArrayExpression) {
            const result = [];
            for (const element of node.elements) {
              if (!element || element.type === AST_NODE_TYPES.SpreadElement) {
                continue;
              }
              result.push(
                ...extractElements(element).map((e) => {
                  if (e.array == null) {
                    return {
                      array: node,
                      node: e.node
                    };
                  }
                  return e;
                })
              );
            }
            return result;
          }
          return [{ node, array: null }];
        }
        return {
          JSXAttribute: verifyAttr,
          AstroTemplateLiteralAttribute: verifyAttr
        };
      }
    });
  }
});

// src/utils/string-literal-parser/tokenizer.ts
function inc(pos, cp) {
  return pos + (cp >= 65536 ? 2 : 1);
}
var CP_BACK_SLASH, CP_BACKTICK, CP_CR, CP_LF, CP_OPENING_BRACE, CP_a, CP_A, CP_n, CP_r, CP_t, CP_b, CP_v, CP_f, CP_u, CP_x, CP_0, CP_7, CP_8, CP_9, Tokenizer;
var init_tokenizer2 = __esm({
  "src/utils/string-literal-parser/tokenizer.ts"() {
    CP_BACK_SLASH = "\\".codePointAt(0);
    CP_BACKTICK = "`".codePointAt(0);
    CP_CR = "\r".codePointAt(0);
    CP_LF = "\n".codePointAt(0);
    CP_OPENING_BRACE = "{".codePointAt(0);
    CP_a = "a".codePointAt(0);
    CP_A = "A".codePointAt(0);
    CP_n = "n".codePointAt(0);
    CP_r = "r".codePointAt(0);
    CP_t = "t".codePointAt(0);
    CP_b = "b".codePointAt(0);
    CP_v = "v".codePointAt(0);
    CP_f = "f".codePointAt(0);
    CP_u = "u".codePointAt(0);
    CP_x = "x".codePointAt(0);
    CP_0 = "0".codePointAt(0);
    CP_7 = "7".codePointAt(0);
    CP_8 = "8".codePointAt(0);
    CP_9 = "9".codePointAt(0);
    Tokenizer = class {
      constructor(source, options) {
        this.source = source;
        this.pos = options.start;
        this.end = options.end ?? null;
        this.ecmaVersion = options.ecmaVersion;
      }
      *parseTokens(quote) {
        const inTemplate = quote === CP_BACKTICK;
        const endIndex = this.end ?? this.source.length;
        while (this.pos < endIndex) {
          const start = this.pos;
          const cp = this.source.codePointAt(start);
          if (cp == null) {
            throw new Error("Unterminated string constant");
          }
          this.pos = inc(start, cp);
          if (cp === quote) break;
          if (cp === CP_BACK_SLASH) {
            const { value, kind } = this.readEscape(inTemplate);
            yield {
              type: "EscapeToken",
              kind,
              value,
              range: [start, this.pos]
            };
          } else if (cp === CP_CR || cp === CP_LF) {
            if (inTemplate) {
              if (cp === CP_CR && this.source.codePointAt(this.pos) === CP_LF) {
                this.pos++;
              }
              yield {
                type: "CharacterToken",
                value: "\n",
                range: [start, this.pos]
              };
            } else {
              throw new Error("Unterminated string constant");
            }
          } else {
            if (this.ecmaVersion >= 2019 && (cp === 8232 || cp === 8233) && !inTemplate) {
              throw new Error("Unterminated string constant");
            }
            yield {
              type: "CharacterToken",
              value: String.fromCodePoint(cp),
              range: [start, this.pos]
            };
          }
        }
      }
      // eslint-disable-next-line complexity -- ignore
      readEscape(inTemplate) {
        const cp = this.source.codePointAt(this.pos);
        if (cp == null) {
          throw new Error("Invalid or unexpected token");
        }
        this.pos = inc(this.pos, cp);
        switch (cp) {
          case CP_n:
            return { value: "\n", kind: "special" };
          case CP_r:
            return { value: "\r", kind: "special" };
          case CP_t:
            return { value: "	", kind: "special" };
          case CP_b:
            return { value: "\b", kind: "special" };
          case CP_v:
            return { value: "\v", kind: "special" };
          case CP_f:
            return { value: "\f", kind: "special" };
          // @ts-expect-error -- falls through
          case CP_CR:
            if (this.source.codePointAt(this.pos) === CP_LF) {
              this.pos++;
            }
          // falls through
          case CP_LF:
            return { value: "", kind: "eol" };
          case CP_x:
            return {
              value: String.fromCodePoint(this.readHex(2)),
              kind: "hex"
            };
          case CP_u:
            return {
              value: String.fromCodePoint(this.readUnicode()),
              kind: "unicode"
            };
          default:
            if (CP_0 <= cp && cp <= CP_7) {
              let octalStr = /^[0-7]+/u.exec(
                this.source.slice(this.pos - 1, this.pos + 2)
              )[0];
              let octal = parseInt(octalStr, 8);
              if (octal > 255) {
                octalStr = octalStr.slice(0, -1);
                octal = parseInt(octalStr, 8);
              }
              this.pos += octalStr.length - 1;
              const nextCp = this.source.codePointAt(this.pos);
              if ((octalStr !== "0" || nextCp === CP_8 || nextCp === CP_9) && inTemplate) {
                throw new Error("Octal literal in template string");
              }
              return {
                value: String.fromCodePoint(octal),
                kind: "octal"
              };
            }
            return {
              value: String.fromCodePoint(cp),
              kind: "char"
            };
        }
      }
      readUnicode() {
        const cp = this.source.codePointAt(this.pos);
        if (cp === CP_OPENING_BRACE) {
          if (this.ecmaVersion < 2015) {
            throw new Error(`Unexpected character '${String.fromCodePoint(cp)}'`);
          }
          this.pos++;
          const endIndex = this.source.indexOf("}", this.pos);
          if (endIndex < 0) {
            throw new Error("Invalid Unicode escape sequence");
          }
          const code = this.readHex(endIndex - this.pos);
          this.pos++;
          if (code > 1114111) {
            throw new Error("Code point out of bounds");
          }
          return code;
        }
        return this.readHex(4);
      }
      readHex(length) {
        let total = 0;
        for (let i = 0; i < length; i++, this.pos++) {
          const cp = this.source.codePointAt(this.pos);
          if (cp == null) {
            throw new Error(`Invalid hexadecimal escape sequence`);
          }
          let val;
          if (CP_a <= cp) {
            val = cp - CP_a + 10;
          } else if (CP_A <= cp) {
            val = cp - CP_A + 10;
          } else if (CP_0 <= cp && cp <= CP_9) {
            val = cp - CP_0;
          } else {
            throw new Error(`Invalid hexadecimal escape sequence`);
          }
          if (val >= 16) {
            throw new Error(`Invalid hexadecimal escape sequence`);
          }
          total = total * 16 + val;
        }
        return total;
      }
    };
  }
});

// src/utils/string-literal-parser/parser.ts
function* parseStringTokens(source, option) {
  const startIndex = option?.start ?? 0;
  const ecmaVersion = option?.ecmaVersion ?? Infinity;
  const tokenizer = new Tokenizer(source, {
    start: startIndex,
    end: option?.end,
    ecmaVersion: ecmaVersion >= 6 && ecmaVersion < 2015 ? ecmaVersion + 2009 : ecmaVersion
  });
  yield* tokenizer.parseTokens();
}
var init_parser = __esm({
  "src/utils/string-literal-parser/parser.ts"() {
    init_tokenizer2();
  }
});

// src/utils/string-literal-parser/tokens.ts
var init_tokens = __esm({
  "src/utils/string-literal-parser/tokens.ts"() {
  }
});

// src/utils/string-literal-parser/index.ts
var init_string_literal_parser = __esm({
  "src/utils/string-literal-parser/index.ts"() {
    init_parser();
    init_tokens();
  }
});
function isStringType(node) {
  if (node.type === AST_NODE_TYPES.Literal) {
    return typeof node.value === "string";
  } else if (node.type === AST_NODE_TYPES.TemplateLiteral) {
    return true;
  } else if (node.type === AST_NODE_TYPES.BinaryExpression) {
    return isStringType(node.left) || isStringType(node.right);
  }
  return isStringCallExpression(node);
}
var prefer_split_class_list_default;
var init_prefer_split_class_list = __esm({
  "src/rules/prefer-split-class-list.ts"() {
    init_utils();
    init_ast_utils();
    init_string_literal_parser();
    init_compat();
    prefer_split_class_list_default = createRule("prefer-split-class-list", {
      meta: {
        docs: {
          description: "require use split array elements in `class:list`",
          category: "Stylistic Issues",
          recommended: false
        },
        schema: [
          {
            type: "object",
            properties: {
              splitLiteral: { type: "boolean" }
            },
            additionalProperties: false
          }
        ],
        messages: {
          uselessClsx: "Using `clsx()` for the `class:list` has no effect.",
          split: "Can split elements with spaces."
        },
        fixable: "code",
        type: "suggestion"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        const splitLiteral = Boolean(context.options[0]?.splitLiteral);
        function shouldReport(state) {
          if (state.isFirstElement) {
            if (state.isLeading) {
              return false;
            }
          }
          if (state.isLastElement) {
            if (state.isTrailing) {
              return false;
            }
          }
          if (splitLiteral) {
            return true;
          }
          return state.isLeading || state.isTrailing;
        }
        function verifyAttr(attr) {
          if (getAttributeName(attr) !== "class:list") {
            return;
          }
          if (!attr.value || attr.value.type !== AST_NODE_TYPES.JSXExpressionContainer || attr.value.expression.type === AST_NODE_TYPES.JSXEmptyExpression) {
            return;
          }
          const expression = attr.value.expression;
          verifyExpression(expression, function* (fixer) {
            if (expression.type === AST_NODE_TYPES.ArrayExpression) {
              return;
            }
            yield fixer.insertTextBeforeRange(expression.range, "[");
            yield fixer.insertTextAfterRange(expression.range, "]");
          });
        }
        function verifyExpression(node, transformArray, call) {
          if (node.type === AST_NODE_TYPES.TemplateLiteral) {
            const first = node.quasis[0];
            const last = node.quasis[node.quasis.length - 1];
            for (const quasi of node.quasis) {
              verifyTemplateElement(quasi, {
                isFirstElement: first === quasi,
                isLastElement: last === quasi,
                transformArray,
                call
              });
            }
          } else if (node.type === AST_NODE_TYPES.BinaryExpression) {
            verifyBinaryExpression(node, transformArray);
          } else if (node.type === AST_NODE_TYPES.ArrayExpression) {
            for (const element of node.elements) {
              if (element) {
                verifyExpression(element, transformArray);
              }
            }
          } else if (node.type === AST_NODE_TYPES.Literal) {
            if (splitLiteral && isStringLiteral(node)) {
              verifyStringLiteral(node, {
                isFirstElement: true,
                isLastElement: true,
                transformArray,
                call
              });
            }
          } else if (node.type === AST_NODE_TYPES.CallExpression) {
            if (node.callee.type === AST_NODE_TYPES.MemberExpression && getPropertyName(node.callee) === "trim") {
              verifyExpression(node.callee.object, transformArray, ".trim()");
            }
          }
        }
        function verifyTemplateElement(node, state) {
          const stringEndOffset = node.tail ? node.range[1] - 1 : node.range[1] - 2;
          let isLeading = true;
          const spaces = [];
          for (const ch of parseStringTokens(sourceCode.text, {
            start: node.range[0] + 1,
            end: stringEndOffset
          })) {
            if (ch.value.trim()) {
              if (spaces.length) {
                if (shouldReport({ ...state, isLeading, isTrailing: false })) {
                  reportRange([
                    spaces[0].range[0],
                    spaces[spaces.length - 1].range[1]
                  ]);
                }
                spaces.length = 0;
              }
              isLeading = false;
            } else {
              spaces.push(ch);
            }
          }
          if (spaces.length) {
            if (shouldReport({ ...state, isLeading, isTrailing: true })) {
              reportRange([spaces[0].range[0], spaces[spaces.length - 1].range[1]]);
            }
            spaces.length = 0;
          }
          function reportRange(range) {
            context.report({
              loc: {
                start: sourceCode.getLocFromIndex(range[0]),
                end: sourceCode.getLocFromIndex(range[1])
              },
              messageId: "split",
              *fix(fixer) {
                yield* state.transformArray(fixer);
                yield fixer.replaceTextRange(range, `\`${state.call || ""},\``);
              }
            });
          }
        }
        function verifyBinaryExpression(node, transformArray) {
          const elements = extractConcatExpressions(node, sourceCode);
          if (elements == null) {
            return;
          }
          const first = elements[0];
          const last = elements[elements.length - 1];
          for (const element of elements) {
            if (isStringLiteral(element)) {
              verifyStringLiteral(element, {
                isFirstElement: first === element,
                isLastElement: last === element,
                transformArray
              });
            }
          }
        }
        function verifyStringLiteral(node, state) {
          const quote = sourceCode.text[node.range[0]];
          let isLeading = true;
          const spaces = [];
          for (const ch of parseStringTokens(sourceCode.text, {
            start: node.range[0] + 1,
            end: node.range[1] - 1
          })) {
            if (ch.value.trim()) {
              if (spaces.length) {
                if (shouldReport({ ...state, isLeading, isTrailing: false })) {
                  reportRange(
                    [spaces[0].range[0], spaces[spaces.length - 1].range[1]],
                    { isLeading, isTrailing: false }
                  );
                }
                spaces.length = 0;
              }
              isLeading = false;
            } else {
              spaces.push(ch);
            }
          }
          if (spaces.length) {
            if (shouldReport({ ...state, isLeading, isTrailing: true })) {
              reportRange(
                [spaces[0].range[0], spaces[spaces.length - 1].range[1]],
                { isLeading, isTrailing: true }
              );
            }
            spaces.length = 0;
          }
          function reportRange(range, spaceState) {
            context.report({
              loc: {
                start: sourceCode.getLocFromIndex(range[0]),
                end: sourceCode.getLocFromIndex(range[1])
              },
              messageId: "split",
              *fix(fixer) {
                yield* state.transformArray(fixer);
                let leftQuote = quote;
                let rightQuote = quote;
                const bin = node.parent;
                if (spaceState.isLeading && bin.right === node && isStringType(bin.left)) {
                  leftQuote = "";
                }
                if (spaceState.isTrailing && bin.left === node && isStringType(bin.right)) {
                  rightQuote = "";
                }
                const replaceRange = [...range];
                if (!leftQuote || !rightQuote) {
                  if (!leftQuote) {
                    replaceRange[0]--;
                  }
                  if (!rightQuote) {
                    replaceRange[1]++;
                  }
                  yield fixer.remove(
                    sourceCode.getTokensBetween(bin.left, bin.right, {
                      includeComments: false,
                      filter: (t) => t.value === bin.operator
                    })[0]
                  );
                }
                yield fixer.replaceTextRange(
                  replaceRange,
                  `${leftQuote}${state.call || ""},${rightQuote}`
                );
              }
            });
          }
        }
        function verifyClsx(clsxCall) {
          if (clsxCall.node.type !== AST_NODE_TYPES.CallExpression) {
            return;
          }
          const callNode = clsxCall.node;
          const parent = callNode.parent;
          if (!parent || parent.type !== AST_NODE_TYPES.JSXExpressionContainer || parent.expression !== callNode) {
            return;
          }
          const parentParent = parent.parent;
          if (!parentParent || parentParent.type !== AST_NODE_TYPES.JSXAttribute || parentParent.value !== parent || getAttributeName(parentParent) !== "class:list") {
            return;
          }
          context.report({
            node: clsxCall.node.callee,
            messageId: "uselessClsx",
            *fix(fixer) {
              const openToken = sourceCode.getTokenAfter(callNode.callee, {
                includeComments: false,
                filter: isOpeningParenToken
              });
              const closeToken = sourceCode.getLastToken(callNode);
              yield fixer.removeRange([callNode.range[0], openToken.range[1]]);
              yield fixer.remove(closeToken);
            }
          });
        }
        return {
          Program(node) {
            const sourceCode2 = getSourceCode(context);
            const referenceTracker = new ReferenceTracker(sourceCode2.getScope(node));
            for (const call of referenceTracker.iterateEsmReferences({
              // https://github.com/lukeed/clsx
              clsx: {
                [ReferenceTracker.CALL]: true
              }
            })) {
              verifyClsx(call);
            }
          },
          JSXAttribute: verifyAttr,
          AstroTemplateLiteralAttribute: verifyAttr
        };
      }
    });
  }
});

// src/utils/fix-tracker.ts
var FixTracker, fix_tracker_default;
var init_fix_tracker = __esm({
  "src/utils/fix-tracker.ts"() {
    init_ast_utils();
    FixTracker = class {
      /**
       * Create a new FixTracker.
       * @param fixer A ruleFixer instance.
       * @param sourceCode A SourceCode object for the current code.
       */
      constructor(fixer, sourceCode) {
        this.fixer = fixer;
        this.sourceCode = sourceCode;
        this.retainedRange = null;
      }
      /**
       * Mark the given range as "retained", meaning that other fixes may not
       * may not modify this region in the same pass.
       * @param range The range to retain.
       * @returns The same RuleFixer, for chained calls.
       */
      retainRange(range) {
        this.retainedRange = range;
        return this;
      }
      /**
       * Given a node, find the function containing it (or the entire program) and
       * mark it as retained, meaning that other fixes may not modify it in this
       * pass. This is useful for avoiding conflicts in fixes that modify control
       * flow.
       * @param node The node to use as a starting point.
       * @returns The same RuleFixer, for chained calls.
       */
      retainEnclosingFunction(node) {
        const functionNode = getUpperFunction(node);
        return this.retainRange(
          functionNode ? functionNode.range : this.sourceCode.ast.range
        );
      }
      /**
       * Given a node or token, find the token before and afterward, and mark that
       * range as retained, meaning that other fixes may not modify it in this
       * pass. This is useful for avoiding conflicts in fixes that make a small
       * change to the code where the AST should not be changed.
       * @param nodeOrToken The node or token to use as a starting
       *      point. The token to the left and right are use in the range.
       * @returns The same RuleFixer, for chained calls.
       */
      retainSurroundingTokens(nodeOrToken) {
        const tokenBefore = this.sourceCode.getTokenBefore(nodeOrToken) || nodeOrToken;
        const tokenAfter = this.sourceCode.getTokenAfter(nodeOrToken) || nodeOrToken;
        return this.retainRange([tokenBefore.range[0], tokenAfter.range[1]]);
      }
      /**
       * Create a fix command that replaces the given range with the given text,
       * accounting for any retained ranges.
       * @param range The range to remove in the fix.
       * @param text The text to insert in place of the range.
       * @returns The fix command.
       */
      replaceTextRange(range, text) {
        let actualRange;
        if (this.retainedRange) {
          actualRange = [
            Math.min(this.retainedRange[0], range[0]),
            Math.max(this.retainedRange[1], range[1])
          ];
        } else {
          actualRange = range;
        }
        return this.fixer.replaceTextRange(
          actualRange,
          this.sourceCode.text.slice(actualRange[0], range[0]) + text + this.sourceCode.text.slice(range[1], actualRange[1])
        );
      }
      /**
       * Create a fix command that removes the given node or token, accounting for
       * any retained ranges.
       * @param nodeOrToken The node or token to remove.
       * @returns The fix command.
       */
      remove(nodeOrToken) {
        return this.replaceTextRange(nodeOrToken.range, "");
      }
    };
    fix_tracker_default = FixTracker;
  }
});
var semi_default;
var init_semi = __esm({
  "src/rules/semi.ts"() {
    init_utils();
    init_compat();
    init_ast_utils();
    init_fix_tracker();
    semi_default = createRule("semi", {
      meta: {
        docs: {
          description: "Require or disallow semicolons instead of ASI",
          category: "Extension Rules",
          recommended: false,
          extensionRule: "semi"
        },
        type: "layout",
        fixable: "code",
        schema: {
          anyOf: [
            {
              type: "array",
              items: [
                {
                  type: "string",
                  enum: ["never"]
                },
                {
                  type: "object",
                  properties: {
                    beforeStatementContinuationChars: {
                      type: "string",
                      enum: ["always", "any", "never"]
                    }
                  },
                  additionalProperties: false
                }
              ],
              minItems: 0,
              maxItems: 2
            },
            {
              type: "array",
              items: [
                {
                  type: "string",
                  enum: ["always"]
                },
                {
                  type: "object",
                  properties: {
                    omitLastInOneLineBlock: { type: "boolean" },
                    omitLastInOneLineClassBody: { type: "boolean" }
                  },
                  additionalProperties: false
                }
              ],
              minItems: 0,
              maxItems: 2
            }
          ]
        },
        messages: {
          missingSemi: "Missing semicolon.",
          extraSemi: "Extra semicolon."
        }
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        const OPT_OUT_PATTERN = /^[(+\-/[`]/u;
        const unsafeClassFieldNames = /* @__PURE__ */ new Set(["get", "set", "static"]);
        const unsafeClassFieldFollowers = /* @__PURE__ */ new Set(["*", "in", "instanceof"]);
        const options = context.options[1];
        const never = context.options[0] === "never";
        const exceptOneLine = Boolean(
          options && "omitLastInOneLineBlock" in options && options.omitLastInOneLineBlock
        );
        const exceptOneLineClassBody = Boolean(
          options && "omitLastInOneLineClassBody" in options && options.omitLastInOneLineClassBody
        );
        const beforeStatementContinuationChars = options && "beforeStatementContinuationChars" in options && options.beforeStatementContinuationChars || "any";
        function report(node, missing = false) {
          const lastToken = sourceCode.getLastToken(node);
          let messageId = "missingSemi";
          let fix, loc;
          if (!missing) {
            loc = {
              start: lastToken.loc.end,
              end: getNextLocation(sourceCode, lastToken.loc.end)
            };
            fix = function(fixer) {
              return fixer.insertTextAfter(lastToken, ";");
            };
          } else {
            messageId = "extraSemi";
            loc = lastToken.loc;
            fix = function(fixer) {
              return new fix_tracker_default(fixer, sourceCode).retainSurroundingTokens(lastToken).remove(lastToken);
            };
          }
          context.report({
            node,
            loc,
            messageId,
            fix
          });
        }
        function isRedundantSemi(semiToken) {
          const nextToken = sourceCode.getTokenAfter(semiToken);
          return !nextToken || isClosingBraceToken(nextToken) || isSemicolonToken(nextToken);
        }
        function isEndOfArrowBlock(lastToken) {
          if (!isClosingBraceToken(lastToken)) return false;
          const node = sourceCode.getNodeByRangeIndex(lastToken.range[0]);
          return node.type === "BlockStatement" && node.parent.type === "ArrowFunctionExpression";
        }
        function maybeClassFieldAsiHazard(node) {
          if (node.type !== "PropertyDefinition") return false;
          const needsNameCheck = !node.computed && node.key.type === "Identifier";
          if (needsNameCheck && "name" in node.key && unsafeClassFieldNames.has(node.key.name)) {
            const isStaticStatic = node.static && node.key.name === "static";
            if (!isStaticStatic && !node.value) return true;
          }
          const followingToken = sourceCode.getTokenAfter(node);
          return unsafeClassFieldFollowers.has(followingToken.value);
        }
        function isOnSameLineWithNextToken(node) {
          const prevToken = sourceCode.getLastToken(node, 1);
          const nextToken = sourceCode.getTokenAfter(node);
          return Boolean(nextToken) && isTokenOnSameLine(prevToken, nextToken);
        }
        function maybeAsiHazardAfter(node) {
          const t = node.type;
          if (t === "DoWhileStatement" || t === "BreakStatement" || t === "ContinueStatement" || t === "DebuggerStatement" || t === "ImportDeclaration" || t === "ExportAllDeclaration")
            return false;
          if (t === "ReturnStatement") return Boolean(node.argument);
          if (t === "ExportNamedDeclaration") return Boolean(node.declaration);
          const lastToken = sourceCode.getLastToken(node, 1);
          if (isEndOfArrowBlock(lastToken)) return false;
          return true;
        }
        function maybeAsiHazardBefore(token) {
          return Boolean(token) && OPT_OUT_PATTERN.test(token.value) && token.value !== "++" && token.value !== "--" && token.value !== "---";
        }
        function canRemoveSemicolon(node) {
          const lastToken = sourceCode.getLastToken(node);
          if (isRedundantSemi(lastToken)) return true;
          if (maybeClassFieldAsiHazard(node)) return false;
          if (isOnSameLineWithNextToken(node)) return false;
          if (node.type !== "PropertyDefinition" && beforeStatementContinuationChars === "never" && !maybeAsiHazardAfter(node))
            return true;
          const nextToken = sourceCode.getTokenAfter(node);
          if (!maybeAsiHazardBefore(nextToken)) return true;
          return false;
        }
        function isLastInOneLinerBlock(node) {
          const parent = node.parent;
          const nextToken = sourceCode.getTokenAfter(node);
          if (!nextToken || nextToken.value !== "}") return false;
          if (parent.type === "BlockStatement")
            return parent.loc.start.line === parent.loc.end.line;
          if (parent.type === "StaticBlock") {
            const openingBrace = sourceCode.getFirstToken(parent, {
              skip: 1
            });
            return openingBrace.loc.start.line === parent.loc.end.line;
          }
          return false;
        }
        function isLastInOneLinerClassBody(node) {
          const parent = node.parent;
          const nextToken = sourceCode.getTokenAfter(node);
          if (!nextToken || nextToken.value !== "}") return false;
          if (parent.type === "ClassBody")
            return parent.loc.start.line === parent.loc.end.line;
          return false;
        }
        function checkForSemicolon(node) {
          const lastToken = sourceCode.getLastToken(node);
          const isSemi = isSemicolonToken(lastToken);
          if (never) {
            const nextToken = sourceCode.getTokenAfter(node);
            if (isSemi && canRemoveSemicolon(node)) report(node, true);
            else if (!isSemi && beforeStatementContinuationChars === "always" && node.type !== "PropertyDefinition" && maybeAsiHazardBefore(nextToken))
              report(node);
          } else {
            const oneLinerBlock = exceptOneLine && isLastInOneLinerBlock(node);
            const oneLinerClassBody = exceptOneLineClassBody && isLastInOneLinerClassBody(node);
            const oneLinerBlockOrClassBody = oneLinerBlock || oneLinerClassBody;
            if (isSemi && oneLinerBlockOrClassBody) report(node, true);
            else if (!isSemi && !oneLinerBlockOrClassBody) report(node);
          }
        }
        function checkForSemicolonForVariableDeclaration(node) {
          const parent = node.parent;
          if ((parent.type !== "ForStatement" || parent.init !== node) && (!/^For(?:In|Of)Statement/u.test(parent.type) || parent.left !== node))
            checkForSemicolon(node);
        }
        return {
          VariableDeclaration: checkForSemicolonForVariableDeclaration,
          ExpressionStatement: checkForSemicolon,
          ReturnStatement: checkForSemicolon,
          ThrowStatement: checkForSemicolon,
          DoWhileStatement: checkForSemicolon,
          DebuggerStatement: checkForSemicolon,
          BreakStatement: checkForSemicolon,
          ContinueStatement: checkForSemicolon,
          ImportDeclaration: checkForSemicolon,
          ExportAllDeclaration: checkForSemicolon,
          ExportNamedDeclaration(node) {
            if (!node.declaration) checkForSemicolon(node);
          },
          ExportDefaultDeclaration(node) {
            if (node.declaration.type === "TSInterfaceDeclaration") return;
            if (!/(?:Class|Function)Declaration/u.test(node.declaration.type))
              checkForSemicolon(node);
          },
          PropertyDefinition: checkForSemicolon,
          TSAbstractPropertyDefinition: checkForSemicolon,
          TSDeclareFunction: checkForSemicolon,
          TSExportAssignment: checkForSemicolon,
          TSImportEqualsDeclaration: checkForSemicolon,
          TSTypeAliasDeclaration: checkForSemicolon,
          TSEmptyBodyFunctionExpression: checkForSemicolon
        };
      }
    });
  }
});

// src/rules/sort-attributes.ts
var sort_attributes_default;
var init_sort_attributes = __esm({
  "src/rules/sort-attributes.ts"() {
    init_utils();
    init_compat();
    sort_attributes_default = createRule("sort-attributes", {
      meta: {
        docs: {
          description: "enforce sorting of attributes",
          category: "Stylistic Issues",
          recommended: false
        },
        schema: [
          {
            type: "object",
            properties: {
              type: { type: "string", enum: ["alphabetical", "line-length"] },
              ignoreCase: { type: "boolean" },
              order: { type: "string", enum: ["asc", "desc"] }
            },
            additionalProperties: false
          }
        ],
        messages: {
          unexpectedAstroAttributesOrder: 'Expected "{{right}}" to come before "{{left}}".'
        },
        fixable: "code",
        type: "suggestion"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        return {
          JSXElement(node) {
            const { openingElement } = node;
            const { attributes } = openingElement;
            if (attributes.length <= 1) {
              return;
            }
            function pairwise(nodes, callback) {
              if (nodes.length > 1) {
                for (let i = 1; i < nodes.length; i++) {
                  const left = nodes.at(i - 1);
                  const right = nodes.at(i);
                  if (left && right) {
                    callback(left, right, i - 1);
                  }
                }
              }
            }
            const compareFunc = context.options[0]?.type === "line-length" ? (a, b) => a.size - b.size : (a, b) => formatName(a.name).localeCompare(formatName(b.name));
            const compare = context.options[0]?.order === "desc" ? (left, right) => compareFunc(right, left) : (left, right) => compareFunc(left, right);
            const parts = attributes.reduce(
              (accumulator, attribute) => {
                if (attribute.type === "JSXSpreadAttribute") {
                  accumulator.push([]);
                  return accumulator;
                }
                const name2 = typeof attribute.name.name === "string" ? attribute.name.name : sourceCode.text.slice(...attribute.name.range);
                accumulator[accumulator.length - 1].push({
                  name: name2,
                  node: attribute,
                  size: attribute.range[1] - attribute.range[0]
                });
                return accumulator;
              },
              [[]]
            );
            for (const nodes of parts) {
              pairwise(nodes, (left, right) => {
                if (compare(left, right) > 0) {
                  context.report({
                    node: left.node,
                    messageId: "unexpectedAstroAttributesOrder",
                    data: {
                      left: left.name,
                      right: right.name
                    },
                    fix(fixer) {
                      return fixer.replaceTextRange(
                        [left.node.range[0], right.node.range[1]],
                        sourceCode.text.slice(...right.node.range) + " ".repeat(right.node.range[0] - left.node.range[1]) + sourceCode.text.slice(...left.node.range)
                      );
                    }
                  });
                }
              });
            }
            function formatName(name2) {
              return context.options[0]?.ignoreCase === false ? name2 : name2.toLowerCase();
            }
          }
        };
      }
    });
  }
});

// src/rules/valid-compile.ts
var valid_compile_default;
var init_valid_compile = __esm({
  "src/rules/valid-compile.ts"() {
    init_utils();
    init_compat();
    valid_compile_default = createRule("valid-compile", {
      meta: {
        docs: {
          description: "disallow warnings when compiling.",
          category: "Possible Errors",
          recommended: true
        },
        schema: [],
        messages: {},
        type: "problem"
      },
      create(context) {
        const sourceCode = getSourceCode(context);
        if (!sourceCode.parserServices.isAstro) {
          return {};
        }
        const diagnostics = sourceCode.parserServices.getAstroResult().diagnostics;
        return {
          Program() {
            for (const { text, code, location, severity } of diagnostics) {
              if (severity === 2) {
                context.report({
                  loc: {
                    start: location,
                    end: location
                  },
                  message: `${text} [${code}]`
                });
              }
            }
          }
        };
      }
    });
  }
});
function requireUserLocal(id) {
  try {
    const cwd = process.cwd();
    const relativeTo = path4.join(cwd, "__placeholder__.js");
    return createRequire(relativeTo)(id);
  } catch {
    return null;
  }
}
var init_require_user = __esm({
  "src/utils/resolve-parser/require-user.ts"() {
  }
});

// src/a11y/load.ts
function getPluginJsxA11y() {
  if (loaded) {
    return pluginJsxA11yCache;
  }
  if (!pluginJsxA11yCache) {
    pluginJsxA11yCache = requireUserLocal("eslint-plugin-jsx-a11y");
  }
  if (!pluginJsxA11yCache) {
    if (typeof __require !== "undefined") {
      try {
        pluginJsxA11yCache = __require("eslint-plugin-jsx-a11y");
      } catch {
        loaded = true;
      }
    }
  }
  return pluginJsxA11yCache || null;
}
var pluginJsxA11yCache, loaded;
var init_load = __esm({
  "src/a11y/load.ts"() {
    init_require_user();
    pluginJsxA11yCache = null;
    loaded = false;
  }
});

// src/a11y/keys.ts
var plugin, a11yRuleKeys, a11yConfigKeys;
var init_keys = __esm({
  "src/a11y/keys.ts"() {
    init_load();
    plugin = getPluginJsxA11y();
    a11yRuleKeys = plugin?.rules ? Object.keys(plugin.rules).filter(
      (s) => !plugin?.rules?.[s]?.meta?.deprecated
    ) : [
      "alt-text",
      "anchor-ambiguous-text",
      "anchor-has-content",
      "anchor-is-valid",
      "aria-activedescendant-has-tabindex",
      "aria-props",
      "aria-proptypes",
      "aria-role",
      "aria-unsupported-elements",
      "autocomplete-valid",
      "click-events-have-key-events",
      "control-has-associated-label",
      "heading-has-content",
      "html-has-lang",
      "iframe-has-title",
      "img-redundant-alt",
      "interactive-supports-focus",
      "label-has-associated-control",
      "lang",
      "media-has-caption",
      "mouse-events-have-key-events",
      "no-access-key",
      "no-aria-hidden-on-focusable",
      "no-autofocus",
      "no-distracting-elements",
      "no-interactive-element-to-noninteractive-role",
      "no-noninteractive-element-interactions",
      "no-noninteractive-element-to-interactive-role",
      "no-noninteractive-tabindex",
      "no-redundant-roles",
      "no-static-element-interactions",
      "prefer-tag-over-role",
      "role-has-required-aria-props",
      "role-supports-aria-props",
      "scope",
      "tabindex-no-positive"
    ];
    a11yConfigKeys = plugin?.configs ? Object.keys(plugin.configs) : ["recommended", "strict"];
  }
});

// src/a11y/configs.ts
function buildFlatConfigs() {
  const configs2 = {};
  for (const configName of a11yConfigKeys) {
    Object.defineProperty(configs2, `jsx-a11y-${configName}`, {
      enumerable: true,
      get() {
        const base = getPluginJsxA11y();
        const baseConfig = base?.configs?.[configName] ?? {};
        const baseRules = baseConfig.rules ?? {};
        const newRules = {};
        for (const ruleName of Object.keys(baseRules)) {
          newRules[`astro/${ruleName}`] = baseRules[ruleName];
        }
        return [
          ...base_default,
          {
            plugins: { "jsx-a11y": base },
            rules: newRules
          }
        ];
      }
    });
  }
  return configs2;
}
var init_configs = __esm({
  "src/a11y/configs.ts"() {
    init_load();
    init_keys();
    init_base();
  }
});

// src/a11y/rules.ts
function getPluginJsxA11yRule(ruleName) {
  const base = getPluginJsxA11y();
  return base?.rules?.[ruleName];
}
function buildRules() {
  const rules4 = [];
  for (const ruleKey of a11yRuleKeys) {
    const jsxRuleName = `jsx-a11y/${ruleKey}`;
    const astroRuleName = `astro/${jsxRuleName}`;
    const ruleWithoutMeta = createRule(jsxRuleName, {
      meta: {
        messages: {},
        schema: [],
        type: "problem",
        docs: {
          description: `apply \`${jsxRuleName}\` rule to Astro components`,
          category: "A11Y Extension Rules",
          recommended: false,
          available: () => Boolean(getPluginJsxA11y())
        }
      },
      create(context) {
        const baseRule = getPluginJsxA11yRule(ruleKey);
        if (!baseRule) {
          context.report({
            loc: { line: 0, column: 0 },
            message: `If you want to use ${astroRuleName} rule, you need to install eslint-plugin-jsx-a11y.`
          });
          return {};
        }
        return defineWrapperListener(baseRule, context);
      }
    });
    const docs = {
      ...ruleWithoutMeta.meta.docs,
      extensionRule: {
        plugin: "eslint-plugin-jsx-a11y",
        get url() {
          return getPluginJsxA11yRule(ruleKey)?.meta?.docs?.url ?? `https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/${ruleKey}.md`;
        }
      }
    };
    const newRule = {
      meta: new Proxy(ruleWithoutMeta.meta, {
        get(_t, key) {
          if (key === "docs") {
            return docs;
          }
          const baseRule = getPluginJsxA11yRule(ruleKey);
          return baseRule?.meta?.[key] ?? ruleWithoutMeta.meta[key];
        }
      }),
      create: ruleWithoutMeta.create
    };
    rules4.push(newRule);
  }
  return rules4;
}
function defineWrapperListener(coreRule, context) {
  const sourceCode = getSourceCode(context);
  if (!sourceCode.parserServices.isAstro) {
    return {};
  }
  const listener = coreRule.create(context);
  const astroListener = {};
  for (const key of Object.keys(listener)) {
    const original = listener[key];
    if (!original) {
      continue;
    }
    const wrappedListener = function(node, ...args) {
      original.call(this, getProxyNode(node), ...args);
    };
    astroListener[key] = wrappedListener;
    const astroKey = key.replace(/(?:^|\b)AstroRawText(?:\b|$)/gu, "JSXText").replace(
      /(?:^|\b)(?:AstroTemplateLiteralAttribute|AstroShorthandAttribute)(?:\b|$)/gu,
      "JSXAttribute"
    );
    if (astroKey !== key) {
      astroListener[astroKey] = wrappedListener;
    }
  }
  function isNode(data) {
    return data && typeof data.type === "string" && Array.isArray(data.range) && data.range.length === 2 && typeof data.range[0] === "number" && typeof data.range[1] === "number";
  }
  function getProxyNode(node, overrides) {
    const type = TYPE_MAP[node.type] || node.type;
    const cache3 = {
      type,
      ...overrides ?? {}
    };
    if (node.type === "JSXAttribute") {
      const attrName = getAttributeName(node);
      const converted = attrName != null && ATTRIBUTE_MAP[attrName];
      if (converted) {
        cache3.name = getProxyNode(node.name, {
          type: "JSXIdentifier",
          namespace: null,
          name: converted
        });
      }
    }
    return new Proxy(node, {
      get(_t, key) {
        if (key in cache3) {
          return cache3[key];
        }
        const data = node[key];
        if (isNode(data)) {
          return cache3[key] = getProxyNode(data);
        }
        if (Array.isArray(data)) {
          return cache3[key] = data.map(
            (e) => isNode(e) ? getProxyNode(e) : e
          );
        }
        return data;
      }
    });
  }
  return astroListener;
}
var TYPE_MAP, ATTRIBUTE_MAP;
var init_rules = __esm({
  "src/a11y/rules.ts"() {
    init_load();
    init_utils();
    init_keys();
    init_ast_utils();
    init_compat();
    TYPE_MAP = {
      AstroRawText: "JSXText",
      AstroTemplateLiteralAttribute: "JSXAttribute",
      AstroShorthandAttribute: "JSXAttribute"
    };
    ATTRIBUTE_MAP = {
      "set:html": "dangerouslySetInnerHTML",
      "set:text": "children",
      autofocus: "autoFocus",
      for: "htmlFor"
    };
  }
});

// src/a11y/index.ts
function buildA11yRules() {
  return buildRules();
}
function buildA11yFlatConfigs() {
  return buildFlatConfigs();
}
var init_a11y = __esm({
  "src/a11y/index.ts"() {
    init_configs();
    init_rules();
  }
});

// src/rules/index.ts
var originalRules, rules;
var init_rules2 = __esm({
  "src/rules/index.ts"() {
    init_missing_client_only_directive_value();
    init_no_conflict_set_directives();
    init_no_deprecated_astro_canonicalurl();
    init_no_deprecated_astro_fetchcontent();
    init_no_deprecated_astro_resolve();
    init_no_deprecated_getentrybyslug();
    init_no_exports_from_components();
    init_no_set_html_directive();
    init_no_set_text_directive();
    init_no_unused_css_selector();
    init_no_unused_define_vars_in_style();
    init_prefer_class_list_directive();
    init_prefer_object_class_list();
    init_prefer_split_class_list();
    init_semi();
    init_sort_attributes();
    init_valid_compile();
    init_a11y();
    originalRules = [
      missing_client_only_directive_value_default,
      no_conflict_set_directives_default,
      no_deprecated_astro_canonicalurl_default,
      no_deprecated_astro_fetchcontent_default,
      no_deprecated_astro_resolve_default,
      no_deprecated_getentrybyslug_default,
      no_exports_from_components_default,
      no_set_html_directive_default,
      no_set_text_directive_default,
      no_unused_css_selector_default,
      no_unused_define_vars_in_style_default,
      prefer_class_list_directive_default,
      prefer_object_class_list_default,
      prefer_split_class_list_default,
      semi_default,
      sort_attributes_default,
      valid_compile_default
    ];
    rules = [...originalRules, ...buildA11yRules()];
  }
});
function isLinterPath(p) {
  return p.includes(`eslint${path4.sep}lib${path4.sep}linter${path4.sep}linter.js`);
}
function getEspree() {
  if (!espreeCache) {
    const linterPath = Object.keys(__require.cache || {}).find(isLinterPath);
    if (linterPath) {
      try {
        espreeCache = createRequire(linterPath)("espree");
      } catch {
      }
    }
  }
  if (!espreeCache) {
    espreeCache = requireUserLocal("espree");
  }
  if (!espreeCache) {
    espreeCache = __require("espree");
  }
  return espreeCache;
}
var espreeCache;
var init_espree = __esm({
  "src/utils/resolve-parser/espree.ts"() {
    init_require_user();
    espreeCache = null;
  }
});

// src/utils/resolve-parser/index.ts
function resolveParser() {
  const modules = [
    "@typescript-eslint/parser",
    "@babel/eslint-parser",
    "espree"
  ];
  for (const id of modules) {
    const parser3 = toParserForESLint(requireUserLocal(id));
    if (!parser3) {
      continue;
    }
    return parser3;
  }
  try {
    return toParserForESLint(__require("@typescript-eslint/parser"));
  } catch {
  }
  return toParserForESLint(getEspree());
}
function toParserForESLint(mod) {
  for (const m of [mod, mod && mod.default]) {
    if (!m) {
      continue;
    }
    if (typeof m.parseForESLint === "function") {
      return m;
    }
    if (typeof m.parse === "function") {
      return {
        parseForESLint(...args) {
          return {
            ast: m.parse(...args)
          };
        }
      };
    }
  }
  return null;
}
var init_resolve_parser = __esm({
  "src/utils/resolve-parser/index.ts"() {
    init_espree();
    init_require_user();
  }
});
function parseExpression(code) {
  const result = resolveParser().parseForESLint(
    `(
${code}
)`,
    { range: true, loc: true }
  );
  const statement = result.ast.body[0];
  const expression = statement.expression;
  traverseNodes(expression, {
    visitorKeys: result.visitorKeys,
    enterNode(node) {
      node.loc.start = {
        ...node.loc.start,
        line: node.loc.start.line - 1
      };
      node.loc.end = {
        ...node.loc.end,
        line: node.loc.end.line - 1
      };
      node.range = [node.range[0] - 2, node.range[1] - 2];
    },
    leaveNode() {
    }
  });
  return expression;
}
var init_parse_expression = __esm({
  "src/shared/client-script/parse-expression.ts"() {
    init_resolve_parser();
  }
});
function getIndent(lines) {
  let indent = null;
  for (const line of lines) {
    if (!line.trim()) continue;
    const lineIndent = (RE_LEADING_SPACES.exec(line) || [""])[0].length;
    if (indent == null) {
      indent = lineIndent;
    } else {
      indent = Math.min(indent, lineIndent);
    }
    if (indent === 0) {
      break;
    }
  }
  return indent || 0;
}
function sortedLastIndex2(array, value) {
  let lower = 0;
  let upper = array.length;
  while (lower < upper) {
    const mid = Math.floor(lower + (upper - lower) / 2);
    const target = array[mid];
    if (target < value) {
      lower = mid + 1;
    } else if (target > value) {
      upper = mid;
    } else {
      return mid + 1;
    }
  }
  return upper;
}
var RE_LEADING_SPACES, seq, Locs, ClientScript;
var init_client_script = __esm({
  "src/shared/client-script/index.ts"() {
    init_parse_expression();
    RE_LEADING_SPACES = /^[\t ]+/u;
    seq = 0;
    Locs = class {
      constructor(lines) {
        const lineStartIndices = [0];
        let index = 0;
        for (const line of lines[lines.length - 1] ? lines : lines.slice(0, -1)) {
          index += line.length;
          lineStartIndices.push(index);
        }
        this.lineStartIndices = lineStartIndices;
      }
      getLocFromIndex(index) {
        const lineNumber = sortedLastIndex2(this.lineStartIndices, index);
        return {
          line: lineNumber,
          column: index - this.lineStartIndices[lineNumber - 1]
        };
      }
    };
    ClientScript = class {
      constructor(code, script, parsed) {
        this.code = code;
        this.script = script;
        this.parsed = parsed;
        this.id = ++seq;
        this.block = this.initBlock();
      }
      initBlock() {
        const textNode = this.script.children[0];
        const startOffset = textNode.position.start.offset;
        const endOffset = this.parsed.getEndOffset(textNode);
        const startLoc = this.parsed.getLocFromIndex(startOffset);
        const lines = this.code.slice(startOffset, endOffset).split(/(?<=\n)/u);
        const firstLine = lines.shift();
        const textLines = [];
        const remapColumnOffsets = [];
        const remapLines = [];
        const defineVars = this.extractDefineVars();
        if (defineVars.length) {
          textLines.push("/* global\n");
          remapLines.push(-1);
          remapColumnOffsets.push(-1);
          for (const defineVar of defineVars) {
            textLines.push(`${defineVar.name}
`);
            remapLines.push(defineVar.loc.line);
            remapColumnOffsets.push(defineVar.loc.column);
          }
          textLines.push("-- define:vars */\n");
          remapLines.push(-1);
          remapColumnOffsets.push(-1);
        }
        if (firstLine.trim()) {
          const firstLineIndent = (RE_LEADING_SPACES.exec(firstLine) || [""])[0].length;
          textLines.push(firstLine.slice(firstLineIndent));
          remapLines.push(startLoc.line);
          remapColumnOffsets.push(firstLineIndent + startLoc.column);
        }
        const indent = getIndent(lines);
        for (let index = 0; index < lines.length; index++) {
          const line = lines[index];
          const lineIndent = Math.min(indent, line.length);
          const lineText = line.slice(lineIndent);
          if (lineText) {
            textLines.push(line.slice(lineIndent));
            remapColumnOffsets.push(lineIndent);
          } else if (line.endsWith("\n")) {
            const eol = line.endsWith("\r\n") ? "\r\n" : "\n";
            textLines.push(eol);
            remapColumnOffsets.push(line.length - eol.length);
          } else {
            textLines.push("");
            remapColumnOffsets.push(lineIndent);
          }
          remapLines.push(startLoc.line + index + 1);
        }
        const text = textLines.join("");
        const textLocs = new Locs(textLines);
        const remapLoc = (loc) => {
          const lineIndex = loc.line - 1;
          if (remapLines.length > lineIndex) {
            return {
              line: remapLines[lineIndex],
              column: loc.column + remapColumnOffsets[lineIndex]
            };
          }
          if (remapLines.length === lineIndex) {
            return this.parsed.getLocFromIndex(endOffset + loc.column);
          }
          return {
            line: -1,
            column: loc.column + 0
          };
        };
        const remapRange = (range) => {
          const startLoc2 = textLocs.getLocFromIndex(range[0]);
          const normalEndLoc = textLocs.getLocFromIndex(range[1]);
          const endLoc = normalEndLoc.column > 0 ? normalEndLoc : textLocs.getLocFromIndex(range[1] - 1);
          const remappedStartLoc = remapLoc(startLoc2);
          const remappedEndLoc = remapLoc(endLoc);
          if (remappedStartLoc.line < 0 || remappedEndLoc.line < 0) {
            return null;
          }
          return [
            this.parsed.getIndexFromLoc(remappedStartLoc),
            this.parsed.getIndexFromLoc(remappedEndLoc) + (normalEndLoc.column > 0 ? 0 : 1)
          ];
        };
        return {
          text,
          remapMessage(message) {
            const loc = remapLoc(message);
            message.line = loc.line;
            message.column = loc.column;
            if (typeof message.endLine === "number" && typeof message.endColumn === "number") {
              const loc2 = remapLoc({
                line: message.endLine,
                column: message.endColumn
              });
              message.endLine = loc2.line;
              message.endColumn = loc2.column;
            }
            if (message.fix) {
              const remappedRange = remapRange(message.fix.range);
              if (remappedRange) {
                message.fix.range = remappedRange;
              } else {
                delete message.fix;
              }
            }
            if (message.suggestions) {
              for (const suggestion of [...message.suggestions]) {
                const remappedRange = remapRange(suggestion.fix.range);
                if (remappedRange) {
                  suggestion.fix.range = remappedRange;
                } else {
                  message.suggestions.splice(
                    message.suggestions.indexOf(suggestion),
                    1
                  );
                }
              }
            }
            return message;
          }
        };
      }
      extractDefineVars() {
        const defineVars = this.script.attributes.find(
          (attr) => attr.kind === "expression" && attr.name === "define:vars"
        );
        if (!defineVars) {
          return [];
        }
        const valueStart = this.parsed.calcAttributeValueStartOffset(defineVars);
        const valueEnd = this.parsed.calcAttributeEndOffset(defineVars);
        let expression;
        try {
          expression = parseExpression(
            this.code.slice(valueStart + 1, valueEnd - 1)
          );
        } catch {
          return [];
        }
        if (expression.type !== AST_NODE_TYPES.ObjectExpression) return [];
        const startLoc = this.parsed.getLocFromIndex(valueStart + 1);
        return expression.properties.filter((p) => p.type === AST_NODE_TYPES.Property).filter((p) => !p.computed).map((p) => {
          return {
            name: p.key.type === AST_NODE_TYPES.Identifier ? p.key.name : p.key.value,
            loc: {
              line: p.key.loc.start.line + startLoc.line - 1,
              column: p.key.loc.start.column + (p.key.loc.start.line === 1 ? startLoc.column : 0)
            }
          };
        });
      }
      getProcessorFile(ext) {
        return {
          text: this.block.text,
          filename: `${this.id}${ext}`
        };
      }
      remapMessages(messages) {
        return messages.filter((m) => !this.isIgnoreMessage(m)).map((m) => this.block.remapMessage(m)).filter((m) => m.line >= 0);
      }
      isIgnoreMessage(message) {
        if ((message.ruleId === "eol-last" || // for test case
        message.ruleId === "rule-to-test/eol-last") && message.messageId === "unexpected") {
          return true;
        }
        return false;
      }
    };
  }
});

// src/shared/index.ts
function beginShared(filename) {
  const result = new Shared();
  sharedMap.set(filename, result);
  return result;
}
function terminateShared(filename) {
  const result = sharedMap.get(filename);
  sharedMap.delete(filename);
  return result ?? null;
}
var Shared, sharedMap;
var init_shared = __esm({
  "src/shared/index.ts"() {
    init_client_script();
    Shared = class {
      constructor() {
        this.clientScripts = [];
      }
      addClientScript(code, node, parsed) {
        const clientScript = new ClientScript(code, node, parsed);
        this.clientScripts.push(clientScript);
        return clientScript;
      }
    };
    sharedMap = /* @__PURE__ */ new Map();
  }
});

// src/meta.ts
var meta_exports = {};
__export(meta_exports, {
  name: () => name,
  version: () => version
});
var name, version;
var init_meta = __esm({
  "src/meta.ts"() {
    name = "eslint-plugin-astro";
    version = "1.3.1";
  }
});
function preprocess(code, filename, virtualFileExt) {
  if (filename) {
    const shared = beginShared(filename);
    let parsed;
    try {
      parsed = parseTemplate(code);
    } catch {
      return [code];
    }
    parsed.walk(parsed.result.ast, (node) => {
      if (node.type === "element" && node.name === "script" && node.children.length && !node.attributes.some(
        ({ name: name2, value }) => name2 === "type" && /json$|importmap/i.test(value)
      )) {
        shared.addClientScript(code, node, parsed);
      }
    });
    return [
      code,
      ...shared.clientScripts.map((cs) => cs.getProcessorFile(virtualFileExt))
    ];
  }
  return [code];
}
function postprocess([messages, ...blockMessages], filename) {
  const shared = terminateShared(filename);
  if (shared) {
    return messages.concat(
      ...blockMessages.map((m, i) => shared.clientScripts[i].remapMessages(m))
    );
  }
  return messages;
}
var astroProcessor, clientSideTsProcessor;
var init_processor = __esm({
  "src/processor/index.ts"() {
    init_shared();
    init_meta();
    astroProcessor = {
      preprocess(code, filename) {
        return preprocess(code, filename, ".js");
      },
      postprocess,
      supportsAutofix: true,
      meta: meta_exports
    };
    clientSideTsProcessor = {
      preprocess(code, filename) {
        return preprocess(code, filename, ".ts");
      },
      postprocess,
      supportsAutofix: true,
      meta: { ...meta_exports, name: "astro/client-side-ts" }
    };
  }
});

// src/plugin.ts
var plugin_exports = {};
__export(plugin_exports, {
  plugin: () => plugin2
});
var rules2, processors, plugin2;
var init_plugin = __esm({
  "src/plugin.ts"() {
    init_rules2();
    init_processor();
    init_meta();
    init_environments();
    rules2 = rules.reduce(
      (obj, r) => {
        obj[r.meta.docs.ruleName] = r;
        return obj;
      },
      {}
    );
    processors = {
      ".astro": astroProcessor,
      astro: astroProcessor,
      "client-side-ts": clientSideTsProcessor
    };
    plugin2 = {
      meta: { name, version },
      environments,
      rules: rules2,
      processors
    };
  }
});
var plugin3, base_default;
var init_base = __esm({
  "src/configs/flat/base.ts"() {
    init_has_typescript_eslint_parser();
    init_environments();
    base_default = [
      {
        name: "astro/base/plugin",
        plugins: {
          get astro() {
            return plugin3 ?? (plugin3 = (init_plugin(), __toCommonJS(plugin_exports)).plugin);
          }
        }
      },
      {
        name: "astro/base",
        files: ["*.astro", "**/*.astro"],
        languageOptions: {
          globals: {
            ...globals.node,
            ...environments.astro.globals
          },
          parser: parser2,
          // The script of Astro components uses ESM.
          sourceType: "module",
          parserOptions: {
            parser: tsESLintParser ?? void 0,
            extraFileExtensions: [".astro"]
          }
        },
        rules: {
          // eslint-plugin-astro rules
          // Enable base rules
        },
        processor: hasTypescriptEslintParser ? "astro/client-side-ts" : "astro/astro"
      },
      {
        // Define the configuration for `<script>` tag.
        // Script in `<script>` is assigned a virtual file name with the `.js` extension.
        name: "astro/base/javascript",
        files: ["**/*.astro/*.js", "*.astro/*.js"],
        languageOptions: {
          globals: {
            ...globals.browser
          },
          sourceType: "module"
        },
        rules: {
          // If you are using "prettier/prettier" rule,
          // you don't need to format inside <script> as it will be formatted as a `.astro` file.
          "prettier/prettier": "off"
        }
      },
      {
        // Define the configuration for `<script>` tag when using `client-side-ts` processor.
        // Script in `<script>` is assigned a virtual file name with the `.ts` extension.
        name: "astro/base/typescript",
        files: ["**/*.astro/*.ts", "*.astro/*.ts"],
        languageOptions: {
          globals: {
            ...globals.browser
          },
          parser: tsESLintParser ?? void 0,
          sourceType: "module",
          parserOptions: {
            project: null
          }
        },
        rules: {
          // If you are using "prettier/prettier" rule,
          // you don't need to format inside <script> as it will be formatted as a `.astro` file.
          "prettier/prettier": "off"
        }
      }
    ];
  }
});

// src/esm-config-builder.ts
init_base();

// src/configs/flat/recommended.ts
init_base();
var recommended_default = [
  ...base_default,
  {
    name: "astro/recommended",
    rules: {
      // eslint-plugin-astro rules
      "astro/missing-client-only-directive-value": "error",
      "astro/no-conflict-set-directives": "error",
      "astro/no-deprecated-astro-canonicalurl": "error",
      "astro/no-deprecated-astro-fetchcontent": "error",
      "astro/no-deprecated-astro-resolve": "error",
      "astro/no-deprecated-getentrybyslug": "error",
      "astro/no-unused-define-vars-in-style": "error",
      "astro/valid-compile": "error"
    }
  }
];

// src/configs/flat/all.ts
init_rules2();
var all = {};
for (const rule of rules.filter(
  (rule2) => rule2.meta.docs.available() && !rule2.meta.deprecated
)) {
  all[rule.meta.docs.ruleId] = "error";
}
var all_default = [
  ...recommended_default,
  {
    rules: {
      ...all,
      ...recommended_default[recommended_default.length - 1].rules
    }
  }
];

// src/esm-config-builder.ts
init_a11y();
function buildEsmConfigs() {
  const esmConfigs = {
    base: base_default,
    recommended: recommended_default,
    all: all_default,
    "jsx-a11y-strict": null,
    "jsx-a11y-recommended": null,
    // For backward compatibility
    "flat/base": base_default,
    "flat/recommended": recommended_default,
    "flat/all": all_default,
    "flat/jsx-a11y-strict": null,
    "flat/jsx-a11y-recommended": null
  };
  const a11yFlatConfigs = buildA11yFlatConfigs();
  for (const configName of Object.keys(a11yFlatConfigs)) {
    Object.defineProperty(esmConfigs, configName, {
      enumerable: true,
      get() {
        return a11yFlatConfigs[configName];
      }
    });
    Object.defineProperty(esmConfigs, `flat/${configName}`, {
      enumerable: true,
      get() {
        return a11yFlatConfigs[configName];
      }
    });
  }
  return esmConfigs;
}

// src/index.mts
init_plugin();
var configs = buildEsmConfigs();
var src_default = Object.assign(plugin2, { configs });
var { meta, rules: rules3, processors: processors2, environments: environments2 } = plugin2;

export { configs, src_default as default, environments2 as environments, meta, processors2 as processors, rules3 as rules };
